<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html><head><title>Python: module pyiomica.pyiomica</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
</head><body bgcolor="#f0f0f8">

<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="heading">
<tr bgcolor="#7799ee">
<td valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial">&nbsp;<br><big><big><strong><a href="pyiomica.html"><font color="#ffffff">pyiomica</font></a>.pyiomica</strong></big></big></font></td
><td align=right valign=bottom
><font color="#ffffff" face="helvetica, arial"><a href=".">index</a><br><a href="https://github.com/gmiaslab/pyiomica/">PyIOmica Github Repository</a></font></td></tr></table>
    <p><tt>PyIOmica&nbsp;is&nbsp;a&nbsp;general&nbsp;omics&nbsp;package&nbsp;with&nbsp;multiple&nbsp;tools&nbsp;for&nbsp;analyzing&nbsp;omics&nbsp;data.<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;from&nbsp;pyiomica&nbsp;import&nbsp;pyiomica<br>
Notes:<br>
&nbsp;&nbsp;&nbsp;&nbsp;For&nbsp;additional&nbsp;information&nbsp;visit:&nbsp;https://github.com/gmiaslab/pyiomica&nbsp;and&nbsp;https://mathiomica.org&nbsp;by&nbsp;G.&nbsp;Mias&nbsp;Lab</tt></p>
<p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#aa55cc">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Modules</strong></big></font></td></tr>
    
<tr><td bgcolor="#aa55cc"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><table width="100%" summary="list"><tr><td width="25%" valign=top><a href="appdirs.html">appdirs</a><br>
<a href="matplotlib.cm.html">matplotlib.cm</a><br>
<a href="copy.html">copy</a><br>
<a href="datetime.html">datetime</a><br>
<a href="gzip.html">gzip</a><br>
<a href="h5py.html">h5py</a><br>
<a href="scipy.cluster.hierarchy.html">scipy.cluster.hierarchy</a><br>
</td><td width="25%" valign=top><a href="json.html">json</a><br>
<a href="matplotlib.html">matplotlib</a><br>
<a href="multiprocessing.html">multiprocessing</a><br>
<a href="numpy.html">numpy</a><br>
<a href="numba.html">numba</a><br>
<a href="networkx.html">networkx</a><br>
<a href="os.html">os</a><br>
</td><td width="25%" valign=top><a href="matplotlib.patheffects.html">matplotlib.patheffects</a><br>
<a href="pandas.html">pandas</a><br>
<a href="pickle.html">pickle</a><br>
<a href="matplotlib.pyplot.html">matplotlib.pyplot</a><br>
<a href="pymysql.html">pymysql</a><br>
<a href="importlib.resources.html">importlib.resources</a><br>
<a href="scipy.html">scipy</a><br>
</td><td width="25%" valign=top><a href="shutil.html">shutil</a><br>
<a href="sklearn.html">sklearn</a><br>
<a href="urllib.html">urllib</a><br>
<a href="zipfile.html">zipfile</a><br>
</td></tr></table></td></tr></table><p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#eeaa77">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Functions</strong></big></font></td></tr>
    
<tr><td bgcolor="#eeaa77"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><dl><dt><a name="-BenjaminiHochbergFDR"><strong>BenjaminiHochbergFDR</strong></a>(pValues, SignificanceLevel=0.05)</dt><dd><tt>HypothesisTesting&nbsp;BenjaminiHochbergFDR&nbsp;correction<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;pValues:&nbsp;p-values&nbsp;(1D&nbsp;array&nbsp;of&nbsp;floats)<br>
&nbsp;&nbsp;&nbsp;&nbsp;SignificanceLevel:&nbsp;default&nbsp;=&nbsp;0.05.<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Corrected&nbsp;p-Values,&nbsp;p-&nbsp;and&nbsp;q-Value&nbsp;cuttoffs<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;result&nbsp;=&nbsp;<a href="#-BenjaminiHochbergFDR">BenjaminiHochbergFDR</a>(pValues)</tt></dd></dl>
 <dl><dt><a name="-ExportEnrichmentReport"><strong>ExportEnrichmentReport</strong></a>(data, AppendString='', OutputDirectory=None)</dt><dd><tt>Export&nbsp;results&nbsp;from&nbsp;enrichment&nbsp;analysis&nbsp;to&nbsp;Excel&nbsp;spreadsheets.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;data:&nbsp;enrichment&nbsp;results<br>
&nbsp;&nbsp;&nbsp;&nbsp;AppendString:&nbsp;custom&nbsp;report&nbsp;name,&nbsp;if&nbsp;empty&nbsp;then&nbsp;time&nbsp;stamp&nbsp;will&nbsp;be&nbsp;used<br>
&nbsp;&nbsp;&nbsp;&nbsp;OutputDirectory:&nbsp;path&nbsp;of&nbsp;directories&nbsp;where&nbsp;the&nbsp;report&nbsp;will&nbsp;be&nbsp;saved<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;None<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#-ExportEnrichmentReport">ExportEnrichmentReport</a>(goExample1,&nbsp;AppendString='goExample1',&nbsp;OutputDirectory=None)</tt></dd></dl>
 <dl><dt><a name="-GOAnalysis"><strong>GOAnalysis</strong></a>(data, GetGeneDictionaryOptions={}, AugmentDictionary=True, InputID=['UniProt ID', 'Gene Symbol'], OutputID='UniProt ID', GOAnalysisAssignerOptions={}, BackgroundSet=[], Species='human', OntologyLengthFilter=2, ReportFilter=1, ReportFilterFunction=&lt;ufunc 'greater_equal'&gt;, pValueCutoff=0.05, TestFunction=&lt;function &lt;lambda&gt; at 0x7fad40c80bf8&gt;, HypothesisFunction=&lt;function &lt;lambda&gt; at 0x7fad40c80c80&gt;, FilterSignificant=True, OBODictionaryVariable=None, OBOGODictionaryOptions={}, MultipleListCorrection=None, MultipleList=False, GeneDictionary=None)</dt><dd><tt>Calculate&nbsp;input&nbsp;data&nbsp;over-representation&nbsp;analysis&nbsp;for&nbsp;Gene&nbsp;Ontology&nbsp;(GO)&nbsp;categories.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;data:&nbsp;data&nbsp;to&nbsp;analyze<br>
&nbsp;&nbsp;&nbsp;&nbsp;GetGeneDictionaryOptions:&nbsp;a&nbsp;list&nbsp;of&nbsp;options&nbsp;that&nbsp;will&nbsp;be&nbsp;passed&nbsp;to&nbsp;this&nbsp;internal&nbsp;GetGeneDictionary&nbsp;function<br>
&nbsp;&nbsp;&nbsp;&nbsp;AugmentDictionary:&nbsp;a&nbsp;choice&nbsp;whether&nbsp;or&nbsp;not&nbsp;to&nbsp;augment&nbsp;the&nbsp;current&nbsp;ConstantGeneDictionary&nbsp;global&nbsp;variable&nbsp;or&nbsp;create&nbsp;a&nbsp;new&nbsp;one<br>
&nbsp;&nbsp;&nbsp;&nbsp;InputID:&nbsp;&nbsp;kind&nbsp;of&nbsp;identifiers/accessions&nbsp;used&nbsp;as&nbsp;input<br>
&nbsp;&nbsp;&nbsp;&nbsp;OutputID:&nbsp;kind&nbsp;of&nbsp;IDs/accessions&nbsp;to&nbsp;convert&nbsp;the&nbsp;input&nbsp;IDs/accession&nbsp;numbers&nbsp;in&nbsp;the&nbsp;function's&nbsp;analysis<br>
&nbsp;&nbsp;&nbsp;&nbsp;GOAnalysisAssignerOptions:&nbsp;a&nbsp;list&nbsp;of&nbsp;options&nbsp;that&nbsp;will&nbsp;be&nbsp;passed&nbsp;to&nbsp;the&nbsp;internal&nbsp;GOAnalysisAssigner&nbsp;function<br>
&nbsp;&nbsp;&nbsp;&nbsp;BackgroundSet:&nbsp;background&nbsp;list&nbsp;to&nbsp;create&nbsp;annotation&nbsp;projection&nbsp;to&nbsp;limited&nbsp;background&nbsp;space,&nbsp;involves<br>
considering&nbsp;pathways/groups/sets&nbsp;and&nbsp;that&nbsp;provides&nbsp;a&nbsp;list&nbsp;of&nbsp;IDs&nbsp;(e.g.&nbsp;gene&nbsp;accessions)&nbsp;that&nbsp;should&nbsp;be&nbsp;<br>
considered&nbsp;as&nbsp;the&nbsp;background&nbsp;for&nbsp;the&nbsp;calculation<br>
&nbsp;&nbsp;&nbsp;&nbsp;Species:&nbsp;the&nbsp;species&nbsp;considered&nbsp;in&nbsp;the&nbsp;calculation,&nbsp;by&nbsp;default&nbsp;corresponding&nbsp;to&nbsp;human<br>
&nbsp;&nbsp;&nbsp;&nbsp;OntologyLengthFilter:&nbsp;function&nbsp;that&nbsp;can&nbsp;be&nbsp;used&nbsp;to&nbsp;set&nbsp;the&nbsp;value&nbsp;for&nbsp;which&nbsp;terms&nbsp;to&nbsp;consider&nbsp;in&nbsp;the&nbsp;computation,&nbsp;<br>
by&nbsp;excluding&nbsp;GO&nbsp;terms&nbsp;that&nbsp;have&nbsp;fewer&nbsp;items&nbsp;compared&nbsp;to&nbsp;the&nbsp;OntologyLengthFilter&nbsp;value.&nbsp;It&nbsp;is&nbsp;used&nbsp;by&nbsp;the&nbsp;internal<br>
GOAnalysisAssigner&nbsp;function<br>
&nbsp;&nbsp;&nbsp;&nbsp;ReportFilter:&nbsp;functions&nbsp;that&nbsp;use&nbsp;pathways/ontologies/groups,&nbsp;and&nbsp;provides&nbsp;a&nbsp;cutoff&nbsp;for&nbsp;membership&nbsp;in&nbsp;ontologies/pathways/groups<br>
in&nbsp;selecting&nbsp;which&nbsp;terms/categories&nbsp;to&nbsp;return.&nbsp;It&nbsp;is&nbsp;typically&nbsp;used&nbsp;in&nbsp;conjunction&nbsp;with&nbsp;ReportFilterFunction<br>
&nbsp;&nbsp;&nbsp;&nbsp;ReportFilterFunction:&nbsp;specifies&nbsp;what&nbsp;operator&nbsp;form&nbsp;will&nbsp;be&nbsp;used&nbsp;to&nbsp;compare&nbsp;against&nbsp;ReportFilter&nbsp;option&nbsp;value&nbsp;in&nbsp;<br>
selecting&nbsp;which&nbsp;terms/categories&nbsp;to&nbsp;return<br>
&nbsp;&nbsp;&nbsp;&nbsp;HypothesisFunction:&nbsp;allows&nbsp;the&nbsp;choice&nbsp;of&nbsp;function&nbsp;for&nbsp;implementing&nbsp;multiple&nbsp;hypothesis&nbsp;testing&nbsp;considerations<br>
&nbsp;&nbsp;&nbsp;&nbsp;FilterSignificant:&nbsp;can&nbsp;be&nbsp;set&nbsp;to&nbsp;True&nbsp;to&nbsp;filter&nbsp;data&nbsp;based&nbsp;on&nbsp;whether&nbsp;the&nbsp;analysis&nbsp;result&nbsp;is&nbsp;statistically&nbsp;significant,&nbsp;<br>
or&nbsp;if&nbsp;set&nbsp;to&nbsp;False&nbsp;to&nbsp;return&nbsp;all&nbsp;membership&nbsp;computations<br>
&nbsp;&nbsp;&nbsp;&nbsp;OBODictionaryVariable:&nbsp;a&nbsp;GO&nbsp;annotation&nbsp;variable.&nbsp;If&nbsp;set&nbsp;to&nbsp;None,&nbsp;OBOGODictionary&nbsp;will&nbsp;be&nbsp;used&nbsp;internally&nbsp;to&nbsp;<br>
automatically&nbsp;generate&nbsp;the&nbsp;default&nbsp;GO&nbsp;annotation<br>
&nbsp;&nbsp;&nbsp;&nbsp;OBOGODictionaryOptions:&nbsp;a&nbsp;list&nbsp;of&nbsp;options&nbsp;to&nbsp;be&nbsp;passed&nbsp;to&nbsp;the&nbsp;internal&nbsp;OBOGODictionary&nbsp;function&nbsp;that&nbsp;provides&nbsp;the&nbsp;GO&nbsp;annotations<br>
&nbsp;&nbsp;&nbsp;&nbsp;MultipleListCorrection:&nbsp;specifies&nbsp;whether&nbsp;or&nbsp;not&nbsp;to&nbsp;correct&nbsp;for&nbsp;multi-omics&nbsp;analysis.&nbsp;The&nbsp;choices&nbsp;are&nbsp;None,&nbsp;Automatic,&nbsp;<br>
or&nbsp;a&nbsp;custom&nbsp;number,&nbsp;e.g&nbsp;protein+RNA<br>
&nbsp;&nbsp;&nbsp;&nbsp;MultipleList:&nbsp;specifies&nbsp;whether&nbsp;the&nbsp;input&nbsp;accessions&nbsp;list&nbsp;constituted&nbsp;a&nbsp;multi-omics&nbsp;list&nbsp;input&nbsp;that&nbsp;is&nbsp;annotated&nbsp;so<br>
&nbsp;&nbsp;&nbsp;&nbsp;GeneDictionary:&nbsp;points&nbsp;to&nbsp;an&nbsp;existing&nbsp;variable&nbsp;to&nbsp;use&nbsp;as&nbsp;a&nbsp;gene&nbsp;dictionary&nbsp;in&nbsp;annotations.&nbsp;If&nbsp;set&nbsp;to&nbsp;None&nbsp;<br>
the&nbsp;default&nbsp;ConstantGeneDictionary&nbsp;will&nbsp;be&nbsp;used<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Enrichment&nbsp;dictionary<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;goExample1&nbsp;=&nbsp;<a href="#-GOAnalysis">GOAnalysis</a>(["TAB1",&nbsp;"TNFSF13B",&nbsp;"MALT1",&nbsp;"TIRAP",&nbsp;"CHUK",&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"TNFRSF13C",&nbsp;"PARP1",&nbsp;"CSNK2A1",&nbsp;"CSNK2A2",&nbsp;"CSNK2B",&nbsp;"LTBR",&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"LYN",&nbsp;"MYD88",&nbsp;"GADD45B",&nbsp;"ATM",&nbsp;"NFKB1",&nbsp;"NFKB2",&nbsp;"NFKBIA",&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"IRAK4",&nbsp;"PIAS4",&nbsp;"PLAU"])</tt></dd></dl>
 <dl><dt><a name="-GOAnalysisAssigner"><strong>GOAnalysisAssigner</strong></a>(PyIOmicaDataDirectory=None, ImportDirectly=False, BackgroundSet=[], Species='human', LengthFilter=None, LengthFilterFunction=&lt;ufunc 'greater_equal'&gt;, GOFileName=None, GOFileColumns=[2, 5], GOURL='http://current.geneontology.org/annotations/')</dt><dd><tt>Download&nbsp;and&nbsp;create&nbsp;gene&nbsp;associations&nbsp;and&nbsp;restrict&nbsp;to&nbsp;required&nbsp;background&nbsp;set.<br>
&nbsp;<br>
Args:&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;PyIOmicaDataDirectory:&nbsp;the&nbsp;directory&nbsp;where&nbsp;the&nbsp;default&nbsp;package&nbsp;data&nbsp;is&nbsp;stored<br>
&nbsp;&nbsp;&nbsp;&nbsp;ImportDirectly:&nbsp;import&nbsp;from&nbsp;URL&nbsp;regardles&nbsp;is&nbsp;the&nbsp;file&nbsp;already&nbsp;exists<br>
&nbsp;&nbsp;&nbsp;&nbsp;BackgroundSet:&nbsp;background&nbsp;list&nbsp;to&nbsp;create&nbsp;annotation&nbsp;projection&nbsp;to&nbsp;limited&nbsp;background&nbsp;space,&nbsp;involves<br>
considering&nbsp;pathways/groups/sets&nbsp;and&nbsp;that&nbsp;provides&nbsp;a&nbsp;list&nbsp;of&nbsp;IDs&nbsp;(e.g.&nbsp;gene&nbsp;accessions)&nbsp;that&nbsp;should&nbsp;<br>
be&nbsp;considered&nbsp;as&nbsp;the&nbsp;background&nbsp;for&nbsp;the&nbsp;calculation<br>
&nbsp;&nbsp;&nbsp;&nbsp;Species:&nbsp;species&nbsp;considered&nbsp;in&nbsp;the&nbsp;calculation,&nbsp;by&nbsp;default&nbsp;corresponding&nbsp;to&nbsp;human<br>
&nbsp;&nbsp;&nbsp;&nbsp;LengthFilterFunction:&nbsp;performs&nbsp;computations&nbsp;of&nbsp;membership&nbsp;in&nbsp;pathways/ontologies/groups/sets,&nbsp;<br>
that&nbsp;specifies&nbsp;which&nbsp;function&nbsp;to&nbsp;use&nbsp;to&nbsp;filter&nbsp;the&nbsp;number&nbsp;of&nbsp;members&nbsp;a&nbsp;reported&nbsp;category&nbsp;has&nbsp;<br>
compared&nbsp;to&nbsp;the&nbsp;number&nbsp;typically&nbsp;provided&nbsp;by&nbsp;LengthFilter&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;LengthFilter:&nbsp;argument&nbsp;for&nbsp;LengthFilterFunction<br>
&nbsp;&nbsp;&nbsp;&nbsp;GOFileName:&nbsp;the&nbsp;name&nbsp;for&nbsp;the&nbsp;specific&nbsp;GO&nbsp;file&nbsp;to&nbsp;download&nbsp;from&nbsp;the&nbsp;GOURL&nbsp;if&nbsp;option&nbsp;ImportDirectly&nbsp;is&nbsp;set&nbsp;to&nbsp;True<br>
&nbsp;&nbsp;&nbsp;&nbsp;GOFileColumns:&nbsp;columns&nbsp;to&nbsp;use&nbsp;for&nbsp;IDs&nbsp;and&nbsp;GO:accessions&nbsp;respectively&nbsp;from&nbsp;the&nbsp;downloaded&nbsp;GO&nbsp;annotation&nbsp;file,&nbsp;<br>
used&nbsp;when&nbsp;ImportDirectly&nbsp;is&nbsp;set&nbsp;to&nbsp;True&nbsp;to&nbsp;obtain&nbsp;a&nbsp;new&nbsp;GO&nbsp;association&nbsp;file<br>
&nbsp;&nbsp;&nbsp;&nbsp;GOURL:&nbsp;the&nbsp;location&nbsp;(base&nbsp;URL)&nbsp;where&nbsp;the&nbsp;GO&nbsp;association&nbsp;annotation&nbsp;files&nbsp;are&nbsp;downloaded&nbsp;from<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;IDToGO&nbsp;and&nbsp;GOToID&nbsp;dictionary<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;GOassignment&nbsp;=&nbsp;<a href="#-GOAnalysisAssigner">GOAnalysisAssigner</a>()</tt></dd></dl>
 <dl><dt><a name="-GeneTranslation"><strong>GeneTranslation</strong></a>(InputList, TargetIDList, GeneDictionary, InputID=None, Species='human')</dt><dd><tt>Use&nbsp;geneDictionary&nbsp;to&nbsp;convert&nbsp;inputList&nbsp;IDs&nbsp;to&nbsp;different&nbsp;annotations&nbsp;as&nbsp;indicated&nbsp;by&nbsp;targetIDList.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;InputList:&nbsp;list&nbsp;of&nbsp;names<br>
&nbsp;&nbsp;&nbsp;&nbsp;TargetIDList:&nbsp;target&nbsp;ID&nbsp;list<br>
&nbsp;&nbsp;&nbsp;&nbsp;GeneDictionary:&nbsp;an&nbsp;existing&nbsp;variable&nbsp;to&nbsp;use&nbsp;as&nbsp;a&nbsp;gene&nbsp;dictionary&nbsp;in&nbsp;annotations.&nbsp;<br>
If&nbsp;set&nbsp;to&nbsp;None&nbsp;the&nbsp;default&nbsp;ConstantGeneDictionary&nbsp;will&nbsp;be&nbsp;used<br>
&nbsp;&nbsp;&nbsp;&nbsp;InputID:&nbsp;the&nbsp;kind&nbsp;of&nbsp;identifiers/accessions&nbsp;used&nbsp;as&nbsp;input<br>
&nbsp;&nbsp;&nbsp;&nbsp;Species:&nbsp;the&nbsp;species&nbsp;considered&nbsp;in&nbsp;the&nbsp;calculation,&nbsp;by&nbsp;default&nbsp;corresponding&nbsp;to&nbsp;human<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Dictionary<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;GenDict&nbsp;=&nbsp;<a href="#-GeneTranslation">GeneTranslation</a>(data,&nbsp;"UniProt&nbsp;ID",&nbsp;ConstantGeneDictionary,&nbsp;InputID&nbsp;=&nbsp;["UniProt&nbsp;ID","Gene&nbsp;Symbol"],&nbsp;&nbsp;Species&nbsp;=&nbsp;"human")</tt></dd></dl>
 <dl><dt><a name="-GetGeneDictionary"><strong>GetGeneDictionary</strong></a>(geneUCSCTable=None, UCSCSQLString=None, UCSCSQLSelectLabels=None, ImportDirectly=False, Species='human', KEGGUCSCSplit=[True, 'KEGG Gene ID'])</dt><dd><tt>Create&nbsp;an&nbsp;ID/accession&nbsp;dictionary&nbsp;from&nbsp;a&nbsp;UCSC&nbsp;search&nbsp;-&nbsp;typically&nbsp;of&nbsp;gene&nbsp;annotations.<br>
&nbsp;<br>
Args:&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;geneUCSCTable:&nbsp;path&nbsp;to&nbsp;a&nbsp;geneUCSCTable&nbsp;file<br>
&nbsp;&nbsp;&nbsp;&nbsp;UCSCSQLString:&nbsp;an&nbsp;association&nbsp;to&nbsp;be&nbsp;used&nbsp;to&nbsp;obtain&nbsp;data&nbsp;from&nbsp;the&nbsp;UCSC&nbsp;Browser&nbsp;tables.&nbsp;The&nbsp;key&nbsp;of&nbsp;the&nbsp;association&nbsp;must&nbsp;<br>
match&nbsp;the&nbsp;Species&nbsp;option&nbsp;value&nbsp;used&nbsp;(default:&nbsp;human).&nbsp;The&nbsp;value&nbsp;for&nbsp;the&nbsp;species&nbsp;corresponds&nbsp;to&nbsp;the&nbsp;actual&nbsp;MySQL&nbsp;command&nbsp;used<br>
&nbsp;&nbsp;&nbsp;&nbsp;UCSCSQLSelectLabels:&nbsp;an&nbsp;association&nbsp;to&nbsp;be&nbsp;used&nbsp;to&nbsp;assign&nbsp;key&nbsp;labels&nbsp;for&nbsp;the&nbsp;data&nbsp;improted&nbsp;from&nbsp;the&nbsp;UCSC&nbsp;Browser&nbsp;tables.&nbsp;<br>
The&nbsp;key&nbsp;of&nbsp;the&nbsp;association&nbsp;must&nbsp;match&nbsp;the&nbsp;Species&nbsp;option&nbsp;value&nbsp;used&nbsp;(default:&nbsp;human).&nbsp;The&nbsp;value&nbsp;is&nbsp;a&nbsp;multi&nbsp;component&nbsp;string&nbsp;<br>
list&nbsp;corresponding&nbsp;to&nbsp;the&nbsp;matrices&nbsp;in&nbsp;the&nbsp;data&nbsp;file,&nbsp;or&nbsp;the&nbsp;tables&nbsp;used&nbsp;in&nbsp;the&nbsp;MySQL&nbsp;query&nbsp;provided&nbsp;by&nbsp;UCSCSQLString<br>
&nbsp;&nbsp;&nbsp;&nbsp;ImportDirectly:&nbsp;import&nbsp;from&nbsp;URL&nbsp;regardles&nbsp;is&nbsp;the&nbsp;file&nbsp;already&nbsp;exists<br>
&nbsp;&nbsp;&nbsp;&nbsp;Species:&nbsp;species&nbsp;considered&nbsp;in&nbsp;the&nbsp;calculation,&nbsp;by&nbsp;default&nbsp;corresponding&nbsp;to&nbsp;human<br>
&nbsp;&nbsp;&nbsp;&nbsp;KEGGUCSCSplit:&nbsp;a&nbsp;two&nbsp;component&nbsp;list,&nbsp;{True/False,&nbsp;label}.&nbsp;If&nbsp;the&nbsp;first&nbsp;component&nbsp;is&nbsp;set&nbsp;to&nbsp;True&nbsp;the&nbsp;initially&nbsp;imported&nbsp;KEGG&nbsp;IDs,&nbsp;<br>
identified&nbsp;by&nbsp;the&nbsp;second&nbsp;component&nbsp;label,&nbsp;&nbsp;are&nbsp;split&nbsp;on&nbsp;+&nbsp;string&nbsp;to&nbsp;fix&nbsp;nomenclature&nbsp;issues,&nbsp;retaining&nbsp;the&nbsp;string&nbsp;following&nbsp;+<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Dictionary<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;geneDict&nbsp;=&nbsp;<a href="#-GetGeneDictionary">GetGeneDictionary</a>()</tt></dd></dl>
 <dl><dt><a name="-KEGGAnalysis"><strong>KEGGAnalysis</strong></a>(data, AnalysisType='Genomic', GetGeneDictionaryOptions={}, AugmentDictionary=True, InputID=['UniProt ID', 'Gene Symbol'], OutputID='KEGG Gene ID', MolecularInputID=['cpd'], MolecularOutputID='cpd', KEGGAnalysisAssignerOptions={}, BackgroundSet=[], KEGGOrganism='hsa', KEGGMolecular='cpd', KEGGDatabase='pathway', PathwayLengthFilter=2, ReportFilter=1, ReportFilterFunction=&lt;ufunc 'greater_equal'&gt;, pValueCutoff=0.05, TestFunction=&lt;function &lt;lambda&gt; at 0x7fad40c80f28&gt;, HypothesisFunction=&lt;function &lt;lambda&gt; at 0x7fad40c81048&gt;, FilterSignificant=True, KEGGDictionaryVariable=None, KEGGDictionaryOptions={}, MultipleListCorrection=None, MultipleList=False, GeneDictionary=None, Species='human', MolecularSpecies='compound', NonUCSC=False, PyIOmicaDataDirectory=None)</dt><dd><tt>Calculate&nbsp;input&nbsp;data&nbsp;over-representation&nbsp;analysis&nbsp;for&nbsp;KEGG:&nbsp;Kyoto&nbsp;Encyclopedia&nbsp;of&nbsp;Genes&nbsp;and&nbsp;Genomes&nbsp;pathways.<br>
Input&nbsp;can&nbsp;be&nbsp;a&nbsp;list,&nbsp;a&nbsp;dictionary&nbsp;of&nbsp;lists&nbsp;or&nbsp;a&nbsp;clustering&nbsp;object.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;data:&nbsp;data&nbsp;to&nbsp;analyze<br>
&nbsp;&nbsp;&nbsp;&nbsp;AnalysisType:&nbsp;analysis&nbsp;methods&nbsp;that&nbsp;may&nbsp;be&nbsp;used,&nbsp;"Genomic",&nbsp;"Molecular"&nbsp;or&nbsp;"All"<br>
&nbsp;&nbsp;&nbsp;&nbsp;GetGeneDictionaryOptions:&nbsp;a&nbsp;list&nbsp;of&nbsp;options&nbsp;that&nbsp;will&nbsp;be&nbsp;passed&nbsp;to&nbsp;this&nbsp;internal&nbsp;GetGeneDictionary&nbsp;function<br>
&nbsp;&nbsp;&nbsp;&nbsp;AugmentDictionary:&nbsp;a&nbsp;choice&nbsp;whether&nbsp;or&nbsp;not&nbsp;to&nbsp;augment&nbsp;the&nbsp;current&nbsp;ConstantGeneDictionary&nbsp;global&nbsp;variable&nbsp;or&nbsp;create&nbsp;a&nbsp;new&nbsp;one<br>
&nbsp;&nbsp;&nbsp;&nbsp;InputID:&nbsp;the&nbsp;kind&nbsp;of&nbsp;identifiers/accessions&nbsp;used&nbsp;as&nbsp;input<br>
&nbsp;&nbsp;&nbsp;&nbsp;OutputID:&nbsp;a&nbsp;string&nbsp;value&nbsp;that&nbsp;specifies&nbsp;what&nbsp;kind&nbsp;of&nbsp;IDs/accessions&nbsp;to&nbsp;convert&nbsp;the&nbsp;input&nbsp;IDs/accession&nbsp;<br>
numbers&nbsp;in&nbsp;the&nbsp;function's&nbsp;analysis<br>
&nbsp;&nbsp;&nbsp;&nbsp;MolecularInputID:&nbsp;a&nbsp;string&nbsp;list&nbsp;to&nbsp;indicate&nbsp;the&nbsp;kind&nbsp;of&nbsp;ID&nbsp;to&nbsp;use&nbsp;for&nbsp;the&nbsp;input&nbsp;molecule&nbsp;entries<br>
&nbsp;&nbsp;&nbsp;&nbsp;KEGGAnalysisAssignerOptions:&nbsp;a&nbsp;list&nbsp;of&nbsp;options&nbsp;that&nbsp;will&nbsp;be&nbsp;passed&nbsp;to&nbsp;this&nbsp;internal&nbsp;KEGGAnalysisAssigner&nbsp;function<br>
&nbsp;&nbsp;&nbsp;&nbsp;BackgroundSet:&nbsp;a&nbsp;list&nbsp;of&nbsp;IDs&nbsp;(e.g.&nbsp;gene&nbsp;accessions)&nbsp;that&nbsp;should&nbsp;be&nbsp;considered&nbsp;as&nbsp;the&nbsp;background&nbsp;for&nbsp;the&nbsp;calculation<br>
&nbsp;&nbsp;&nbsp;&nbsp;KEGGOrganism:&nbsp;indicates&nbsp;which&nbsp;organism&nbsp;(org)&nbsp;to&nbsp;use&nbsp;for&nbsp;"Genomic"&nbsp;type&nbsp;of&nbsp;analysis&nbsp;(default&nbsp;is&nbsp;human&nbsp;analysis:&nbsp;org="hsa")<br>
&nbsp;&nbsp;&nbsp;&nbsp;KEGGMolecular:&nbsp;which&nbsp;database&nbsp;to&nbsp;use&nbsp;for&nbsp;molecular&nbsp;analysis&nbsp;(default&nbsp;is&nbsp;the&nbsp;compound&nbsp;database:&nbsp;cpd)<br>
&nbsp;&nbsp;&nbsp;&nbsp;KEGGDatabase:&nbsp;KEGG&nbsp;database&nbsp;to&nbsp;use&nbsp;as&nbsp;the&nbsp;target&nbsp;database<br>
&nbsp;&nbsp;&nbsp;&nbsp;PathwayLengthFilter:&nbsp;pathways&nbsp;to&nbsp;consider&nbsp;in&nbsp;the&nbsp;computation,&nbsp;by&nbsp;excluding&nbsp;pathways&nbsp;that&nbsp;have&nbsp;fewer&nbsp;items&nbsp;<br>
compared&nbsp;to&nbsp;the&nbsp;PathwayLengthFilter&nbsp;value<br>
&nbsp;&nbsp;&nbsp;&nbsp;ReportFilter:&nbsp;provides&nbsp;a&nbsp;cutoff&nbsp;for&nbsp;membership&nbsp;in&nbsp;ontologies/pathways/groups&nbsp;in&nbsp;selecting&nbsp;which&nbsp;terms/categories&nbsp;<br>
to&nbsp;return.&nbsp;It&nbsp;is&nbsp;typically&nbsp;used&nbsp;in&nbsp;conjunction&nbsp;with&nbsp;ReportFilterFunction<br>
&nbsp;&nbsp;&nbsp;&nbsp;ReportFilterFunction:&nbsp;operator&nbsp;form&nbsp;will&nbsp;be&nbsp;used&nbsp;to&nbsp;compare&nbsp;against&nbsp;ReportFilter&nbsp;option&nbsp;value&nbsp;in&nbsp;selecting&nbsp;<br>
which&nbsp;terms/categories&nbsp;to&nbsp;return<br>
&nbsp;&nbsp;&nbsp;&nbsp;pValueCutoff:&nbsp;a&nbsp;cutoff&nbsp;p-value&nbsp;for&nbsp;(adjusted)&nbsp;p-values&nbsp;to&nbsp;assess&nbsp;statistical&nbsp;significance&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;TestFunction:&nbsp;a&nbsp;function&nbsp;used&nbsp;to&nbsp;calculate&nbsp;p-values<br>
&nbsp;&nbsp;&nbsp;&nbsp;HypothesisFunction:&nbsp;allows&nbsp;the&nbsp;choice&nbsp;of&nbsp;function&nbsp;for&nbsp;implementing&nbsp;multiple&nbsp;hypothesis&nbsp;testing&nbsp;considerations<br>
&nbsp;&nbsp;&nbsp;&nbsp;FilterSignificant:&nbsp;can&nbsp;be&nbsp;set&nbsp;to&nbsp;True&nbsp;to&nbsp;filter&nbsp;data&nbsp;based&nbsp;on&nbsp;whether&nbsp;the&nbsp;analysis&nbsp;result&nbsp;is&nbsp;statistically&nbsp;significant,&nbsp;<br>
or&nbsp;if&nbsp;set&nbsp;to&nbsp;False&nbsp;to&nbsp;return&nbsp;all&nbsp;membership&nbsp;computations<br>
&nbsp;&nbsp;&nbsp;&nbsp;KEGGDictionaryVariable:&nbsp;KEGG&nbsp;dictionary,&nbsp;and&nbsp;provides&nbsp;a&nbsp;KEGG&nbsp;annotation&nbsp;variable.&nbsp;If&nbsp;set&nbsp;to&nbsp;None,&nbsp;KEGGDictionary&nbsp;<br>
will&nbsp;be&nbsp;used&nbsp;internally&nbsp;to&nbsp;automatically&nbsp;generate&nbsp;the&nbsp;default&nbsp;KEGG&nbsp;annotation<br>
&nbsp;&nbsp;&nbsp;&nbsp;KEGGDictionaryOptions:&nbsp;a&nbsp;list&nbsp;of&nbsp;options&nbsp;to&nbsp;be&nbsp;passed&nbsp;to&nbsp;the&nbsp;internal&nbsp;KEGGDictionary&nbsp;function&nbsp;that&nbsp;provides&nbsp;the&nbsp;KEGG&nbsp;annotations<br>
&nbsp;&nbsp;&nbsp;&nbsp;MultipleListCorrection:&nbsp;specifies&nbsp;whether&nbsp;or&nbsp;not&nbsp;to&nbsp;correct&nbsp;for&nbsp;multi-omics&nbsp;analysis.&nbsp;<br>
The&nbsp;choices&nbsp;are&nbsp;None,&nbsp;Automatic,&nbsp;or&nbsp;a&nbsp;custom&nbsp;number<br>
&nbsp;&nbsp;&nbsp;&nbsp;MultipleList:&nbsp;whether&nbsp;the&nbsp;input&nbsp;accessions&nbsp;list&nbsp;constituted&nbsp;a&nbsp;multi-omics&nbsp;list&nbsp;input&nbsp;that&nbsp;is&nbsp;annotated&nbsp;so<br>
&nbsp;&nbsp;&nbsp;&nbsp;GeneDictionary:&nbsp;existing&nbsp;variable&nbsp;to&nbsp;use&nbsp;as&nbsp;a&nbsp;gene&nbsp;dictionary&nbsp;in&nbsp;annotations.&nbsp;If&nbsp;set&nbsp;to&nbsp;None&nbsp;the&nbsp;default&nbsp;ConstantGeneDictionary&nbsp;will&nbsp;be&nbsp;used<br>
&nbsp;&nbsp;&nbsp;&nbsp;Species:&nbsp;the&nbsp;species&nbsp;considered&nbsp;in&nbsp;the&nbsp;calculation,&nbsp;by&nbsp;default&nbsp;corresponding&nbsp;to&nbsp;human<br>
&nbsp;&nbsp;&nbsp;&nbsp;MolecularSpecies:&nbsp;the&nbsp;kind&nbsp;of&nbsp;molecular&nbsp;input<br>
&nbsp;&nbsp;&nbsp;&nbsp;NonUCSC:&nbsp;if&nbsp;UCSC&nbsp;browser&nbsp;was&nbsp;used&nbsp;in&nbsp;determining&nbsp;an&nbsp;internal&nbsp;GeneDictionary&nbsp;used&nbsp;in&nbsp;ID&nbsp;translations,<br>
where&nbsp;the&nbsp;KEGG&nbsp;identifiers&nbsp;for&nbsp;genes&nbsp;are&nbsp;number&nbsp;strings&nbsp;(e.g.&nbsp;4790).The&nbsp;NonUCSC&nbsp;option&nbsp;can&nbsp;be&nbsp;set&nbsp;to&nbsp;True&nbsp;<br>
if&nbsp;standard&nbsp;KEGG&nbsp;accessions&nbsp;are&nbsp;used&nbsp;in&nbsp;a&nbsp;user&nbsp;provided&nbsp;GeneDictionary&nbsp;variable,&nbsp;<br>
in&nbsp;the&nbsp;form&nbsp;OptionValue[KEGGOrganism]&nbsp;&lt;&gt;:&lt;&gt;numberString,&nbsp;e.g.&nbsp;hsa:4790<br>
&nbsp;&nbsp;&nbsp;&nbsp;PyIOmicaDataDirectory:&nbsp;directory&nbsp;where&nbsp;the&nbsp;default&nbsp;package&nbsp;data&nbsp;is&nbsp;stored<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Enrichment&nbsp;dictionary<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;keggExample1&nbsp;=&nbsp;<a href="#-KEGGAnalysis">KEGGAnalysis</a>(["TAB1",&nbsp;"TNFSF13B",&nbsp;"MALT1",&nbsp;"TIRAP",&nbsp;"CHUK",&nbsp;"TNFRSF13C",&nbsp;"PARP1",&nbsp;"CSNK2A1",&nbsp;"CSNK2A2",&nbsp;"CSNK2B",&nbsp;"LTBR",&nbsp;"LYN",&nbsp;"MYD88",&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"GADD45B",&nbsp;"ATM",&nbsp;"NFKB1",&nbsp;"NFKB2",&nbsp;"NFKBIA",&nbsp;"IRAK4",&nbsp;"PIAS4",&nbsp;"PLAU",&nbsp;"POLR3B",&nbsp;"NME1",&nbsp;"CTPS1",&nbsp;"POLR3A"])</tt></dd></dl>
 <dl><dt><a name="-KEGGAnalysisAssigner"><strong>KEGGAnalysisAssigner</strong></a>(PyIOmicaDataDirectory=None, ImportDirectly=False, BackgroundSet=[], KEGGQuery1='pathway', KEGGQuery2='hsa', LengthFilter=None, LengthFilterFunction=&lt;ufunc 'greater_equal'&gt;, Labels=['IDToPath', 'PathToID'])</dt><dd><tt>Create&nbsp;KEGG:&nbsp;Kyoto&nbsp;Encyclopedia&nbsp;of&nbsp;Genes&nbsp;and&nbsp;Genomes&nbsp;pathway&nbsp;associations,&nbsp;<br>
restricted&nbsp;to&nbsp;required&nbsp;background&nbsp;set,&nbsp;downloading&nbsp;the&nbsp;data&nbsp;if&nbsp;necessary.<br>
&nbsp;<br>
Args:&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;PyIOmicaDataDirectory:&nbsp;directory&nbsp;where&nbsp;the&nbsp;default&nbsp;package&nbsp;data&nbsp;is&nbsp;stored<br>
&nbsp;&nbsp;&nbsp;&nbsp;ImportDirectly:&nbsp;import&nbsp;from&nbsp;URL&nbsp;regardles&nbsp;is&nbsp;the&nbsp;file&nbsp;already&nbsp;exists<br>
&nbsp;&nbsp;&nbsp;&nbsp;BackgroundSet:&nbsp;a&nbsp;list&nbsp;of&nbsp;IDs&nbsp;(e.g.&nbsp;gene&nbsp;accessions)&nbsp;that&nbsp;should&nbsp;be&nbsp;considered&nbsp;as&nbsp;the&nbsp;background&nbsp;for&nbsp;the&nbsp;calculation<br>
&nbsp;&nbsp;&nbsp;&nbsp;KEGGQuery1:&nbsp;make&nbsp;KEGG&nbsp;API&nbsp;calls,&nbsp;and&nbsp;sets&nbsp;string&nbsp;query1&nbsp;in&nbsp;<a href="http://rest.kegg.jp/link/">http://rest.kegg.jp/link/</a>&lt;&gt;&nbsp;query1&nbsp;&lt;&gt;&nbsp;/&nbsp;&lt;&gt;&nbsp;query2.&nbsp;<br>
Typically&nbsp;this&nbsp;will&nbsp;be&nbsp;used&nbsp;as&nbsp;the&nbsp;target&nbsp;database&nbsp;to&nbsp;find&nbsp;related&nbsp;entries&nbsp;by&nbsp;using&nbsp;database&nbsp;cross-references<br>
&nbsp;&nbsp;&nbsp;&nbsp;KEGGQuery2:&nbsp;KEGG&nbsp;API&nbsp;calls,&nbsp;and&nbsp;sets&nbsp;string&nbsp;query2&nbsp;in&nbsp;<a href="http://rest.kegg.jp/link/">http://rest.kegg.jp/link/</a>&lt;&gt;&nbsp;query1&nbsp;&lt;&gt;&nbsp;/&nbsp;&lt;&gt;&nbsp;query2.&nbsp;<br>
Typically&nbsp;this&nbsp;will&nbsp;be&nbsp;used&nbsp;as&nbsp;the&nbsp;source&nbsp;database&nbsp;to&nbsp;find&nbsp;related&nbsp;entries&nbsp;by&nbsp;using&nbsp;database&nbsp;cross-references<br>
&nbsp;&nbsp;&nbsp;&nbsp;LengthFilterFunction:&nbsp;option&nbsp;for&nbsp;functions&nbsp;that&nbsp;perform&nbsp;computations&nbsp;of&nbsp;membership&nbsp;in&nbsp;<br>
pathways/ontologies/groups/sets,&nbsp;that&nbsp;specifies&nbsp;which&nbsp;function&nbsp;to&nbsp;use&nbsp;to&nbsp;filter&nbsp;the&nbsp;number&nbsp;of&nbsp;members&nbsp;a&nbsp;reported&nbsp;<br>
category&nbsp;has&nbsp;compared&nbsp;to&nbsp;the&nbsp;number&nbsp;typically&nbsp;provided&nbsp;by&nbsp;LengthFilter<br>
&nbsp;&nbsp;&nbsp;&nbsp;LengthFilter:&nbsp;allows&nbsp;the&nbsp;selection&nbsp;of&nbsp;how&nbsp;many&nbsp;members&nbsp;each&nbsp;category&nbsp;can&nbsp;have,&nbsp;as&nbsp;typically&nbsp;<br>
restricted&nbsp;by&nbsp;the&nbsp;LengthFilterFunction<br>
&nbsp;&nbsp;&nbsp;&nbsp;Labels:&nbsp;a&nbsp;string&nbsp;list&nbsp;for&nbsp;how&nbsp;keys&nbsp;in&nbsp;a&nbsp;created&nbsp;association&nbsp;will&nbsp;be&nbsp;named<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;IDToPath&nbsp;and&nbsp;PathToID&nbsp;dictionary<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;KEGGassignment&nbsp;=&nbsp;<a href="#-KEGGAnalysisAssigner">KEGGAnalysisAssigner</a>()</tt></dd></dl>
 <dl><dt><a name="-KEGGDictionary"><strong>KEGGDictionary</strong></a>(PyIOmicaDataDirectory=None, ImportDirectly=False, KEGGQuery1='pathway', KEGGQuery2='hsa')</dt><dd><tt>Create&nbsp;a&nbsp;dictionary&nbsp;from&nbsp;KEGG:&nbsp;Kyoto&nbsp;Encyclopedia&nbsp;of&nbsp;Genes&nbsp;and&nbsp;Genomes&nbsp;terms&nbsp;-&nbsp;<br>
typically&nbsp;association&nbsp;of&nbsp;pathways&nbsp;and&nbsp;members&nbsp;therein.<br>
&nbsp;<br>
Args:&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;PyIOmicaDataDirectory:&nbsp;directory&nbsp;where&nbsp;the&nbsp;default&nbsp;package&nbsp;data&nbsp;is&nbsp;stored<br>
&nbsp;&nbsp;&nbsp;&nbsp;ImportDirectly:&nbsp;import&nbsp;from&nbsp;URL&nbsp;regardles&nbsp;is&nbsp;the&nbsp;file&nbsp;already&nbsp;exists<br>
&nbsp;&nbsp;&nbsp;&nbsp;KEGGQuery1:&nbsp;make&nbsp;KEGG&nbsp;API&nbsp;calls,&nbsp;and&nbsp;sets&nbsp;string&nbsp;query1&nbsp;in&nbsp;<a href="http://rest.kegg.jp/link/">http://rest.kegg.jp/link/</a>&lt;&gt;&nbsp;query1&nbsp;&lt;&gt;&nbsp;/&nbsp;&lt;&gt;&nbsp;query2.&nbsp;<br>
Typically&nbsp;this&nbsp;will&nbsp;be&nbsp;used&nbsp;as&nbsp;the&nbsp;target&nbsp;database&nbsp;to&nbsp;find&nbsp;related&nbsp;entries&nbsp;by&nbsp;using&nbsp;database&nbsp;cross-references<br>
&nbsp;&nbsp;&nbsp;&nbsp;KEGGQuery2:&nbsp;KEGG&nbsp;API&nbsp;calls,&nbsp;and&nbsp;sets&nbsp;string&nbsp;query2&nbsp;in&nbsp;<a href="http://rest.kegg.jp/link/">http://rest.kegg.jp/link/</a>&lt;&gt;&nbsp;query1&nbsp;&lt;&gt;&nbsp;/&nbsp;&lt;&gt;&nbsp;query2.&nbsp;<br>
Typically&nbsp;this&nbsp;will&nbsp;be&nbsp;used&nbsp;as&nbsp;the&nbsp;source&nbsp;database&nbsp;to&nbsp;find&nbsp;related&nbsp;entries&nbsp;by&nbsp;using&nbsp;database&nbsp;cross-references<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Dictionary&nbsp;of&nbsp;definitions<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;KEGGDict&nbsp;=&nbsp;<a href="#-KEGGDictionary">KEGGDictionary</a>()</tt></dd></dl>
 <dl><dt><a name="-LombScargle"><strong>LombScargle</strong></a>(inputTimes, inputData, inputSetTimes, FrequenciesOnly=False, NormalizeIntensities=False, OversamplingRate=1, UpperFrequencyFactor=1)</dt><dd><tt>Calculate&nbsp;Lomb-Scargle&nbsp;periodogram.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;inputTimes:&nbsp;times&nbsp;corresponding&nbsp;to&nbsp;provided&nbsp;data&nbsp;points&nbsp;(1D&nbsp;array&nbsp;of&nbsp;floats)<br>
&nbsp;&nbsp;&nbsp;&nbsp;inputData:&nbsp;data&nbsp;points&nbsp;(1D&nbsp;array&nbsp;of&nbsp;floats)<br>
&nbsp;&nbsp;&nbsp;&nbsp;inputSetTimes:&nbsp;a&nbsp;complete&nbsp;set&nbsp;of&nbsp;all&nbsp;possible&nbsp;N&nbsp;times&nbsp;during&nbsp;which&nbsp;data&nbsp;could&nbsp;have&nbsp;been&nbsp;collected<br>
&nbsp;&nbsp;&nbsp;&nbsp;FrequenciesOnly:&nbsp;return&nbsp;frequencies&nbsp;only<br>
&nbsp;&nbsp;&nbsp;&nbsp;NormalizeIntensities:&nbsp;normalize&nbsp;intensities&nbsp;to&nbsp;unity<br>
&nbsp;&nbsp;&nbsp;&nbsp;OversamplingRate:&nbsp;oversampling&nbsp;rate<br>
&nbsp;&nbsp;&nbsp;&nbsp;UpperFrequencyFactor:&nbsp;upper&nbsp;frequency&nbsp;factor<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Periodogram&nbsp;with&nbsp;a&nbsp;list&nbsp;of&nbsp;frequencies.<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;pgram&nbsp;=&nbsp;<a href="#-LombScargle">LombScargle</a>(inputTimes,&nbsp;inputData,&nbsp;inputSetTimes)</tt></dd></dl>
 <dl><dt><a name="-MassDictionary"><strong>MassDictionary</strong></a>(PyIOmicaDataDirectory=None)</dt><dd><tt>Load&nbsp;PyIOmica's&nbsp;current&nbsp;mass&nbsp;dictionary.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;PyIOmicaDataDirectory:&nbsp;directory&nbsp;where&nbsp;the&nbsp;default&nbsp;package&nbsp;data&nbsp;is&nbsp;stored<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Mass&nbsp;dictionary<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;MassDict&nbsp;=&nbsp;<a href="#-MassDictionary">MassDictionary</a>()</tt></dd></dl>
 <dl><dt><a name="-MassMatcher"><strong>MassMatcher</strong></a>(data, accuracy, MassDictionaryVariable=None, MolecularSpecies='cpd')</dt><dd><tt>Assign&nbsp;putative&nbsp;mass&nbsp;identification&nbsp;to&nbsp;input&nbsp;data&nbsp;based&nbsp;on&nbsp;monoisotopic&nbsp;mass&nbsp;<br>
(using&nbsp;PyIOmica's&nbsp;mass&nbsp;dictionary).&nbsp;The&nbsp;accuracy&nbsp;in&nbsp;parts&nbsp;per&nbsp;million.&nbsp;<br>
&nbsp;<br>
Args:&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;data:&nbsp;input&nbsp;data<br>
&nbsp;&nbsp;&nbsp;&nbsp;accuracy:&nbsp;accuracy<br>
&nbsp;&nbsp;&nbsp;&nbsp;MassDictionaryVariable:&nbsp;mass&nbsp;dictionary&nbsp;variable.&nbsp;If&nbsp;set&nbsp;to&nbsp;None,&nbsp;inbuilt&nbsp;<br>
mass&nbsp;dictionary&nbsp;(MassDictionary)&nbsp;will&nbsp;be&nbsp;loaded&nbsp;and&nbsp;used<br>
&nbsp;&nbsp;&nbsp;&nbsp;MolecularSpecies:&nbsp;the&nbsp;kind&nbsp;of&nbsp;molecular&nbsp;input<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;List&nbsp;of&nbsp;IDs&nbsp;<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;result&nbsp;=&nbsp;<a href="#-MassMatcher">MassMatcher</a>(18.010565,&nbsp;2)</tt></dd></dl>
 <dl><dt><a name="-OBOGODictionary"><strong>OBOGODictionary</strong></a>(FileURL='http://purl.obolibrary.org/obo/go/go-basic.obo', ImportDirectly=False, PyIOmicaDataDirectory=None, OBOFile='goBasicObo.txt')</dt><dd><tt>Generate&nbsp;Open&nbsp;Biomedical&nbsp;Ontologies&nbsp;(OBO)&nbsp;Gene&nbsp;Ontology&nbsp;(GO)&nbsp;vocabulary&nbsp;dictionary.<br>
&nbsp;<br>
Args:&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;FileURL:&nbsp;provides&nbsp;the&nbsp;location&nbsp;of&nbsp;the&nbsp;Open&nbsp;Biomedical&nbsp;Ontologies&nbsp;(OBO)&nbsp;Gene&nbsp;Ontology&nbsp;(GO)&nbsp;<br>
file&nbsp;in&nbsp;case&nbsp;this&nbsp;will&nbsp;be&nbsp;downloaded&nbsp;from&nbsp;the&nbsp;web<br>
&nbsp;&nbsp;&nbsp;&nbsp;ImportDirectly:&nbsp;import&nbsp;from&nbsp;URL&nbsp;regardles&nbsp;is&nbsp;the&nbsp;file&nbsp;already&nbsp;exists<br>
&nbsp;&nbsp;&nbsp;&nbsp;PyIOmicaDataDirectory:&nbsp;path&nbsp;of&nbsp;directories&nbsp;to&nbsp;data&nbsp;storage<br>
&nbsp;&nbsp;&nbsp;&nbsp;OBOFile:&nbsp;name&nbsp;of&nbsp;file&nbsp;to&nbsp;store&nbsp;data&nbsp;in&nbsp;(file&nbsp;will&nbsp;be&nbsp;zipped)<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Dictionary&nbsp;of&nbsp;definitions<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;OBODict&nbsp;=&nbsp;<a href="#-OBOGODictionary">OBOGODictionary</a>()</tt></dd></dl>
 <dl><dt><a name="-PlotHorizontalVisibilityGraph"><strong>PlotHorizontalVisibilityGraph</strong></a>(A, data, times, fileName, id)</dt><dd><tt>Bar-plot&nbsp;style&nbsp;horizontal&nbsp;visibility&nbsp;graph.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;A:&nbsp;Adjacency&nbsp;matrix<br>
&nbsp;&nbsp;&nbsp;&nbsp;data:&nbsp;Numpy&nbsp;2-D&nbsp;array&nbsp;of&nbsp;floats<br>
&nbsp;&nbsp;&nbsp;&nbsp;times:&nbsp;Numpy&nbsp;1-D&nbsp;array&nbsp;of&nbsp;floats<br>
&nbsp;&nbsp;&nbsp;&nbsp;fileName:&nbsp;name&nbsp;of&nbsp;the&nbsp;figure&nbsp;file&nbsp;to&nbsp;save<br>
&nbsp;&nbsp;&nbsp;&nbsp;id:&nbsp;label&nbsp;to&nbsp;add&nbsp;to&nbsp;the&nbsp;figure&nbsp;title<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;None<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#-PlotHorizontalVisibilityGraph">PlotHorizontalVisibilityGraph</a>(A,&nbsp;data,&nbsp;times,&nbsp;'FIgure.png',&nbsp;'Test&nbsp;Data')</tt></dd></dl>
 <dl><dt><a name="-PlotVisibilityGraph"><strong>PlotVisibilityGraph</strong></a>(A, data, times, fileName, id)</dt><dd><tt>Bar-plot&nbsp;style&nbsp;visibility&nbsp;graph.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;A:&nbsp;Adjacency&nbsp;matrix<br>
&nbsp;&nbsp;&nbsp;&nbsp;data:&nbsp;Numpy&nbsp;2-D&nbsp;array&nbsp;of&nbsp;floats<br>
&nbsp;&nbsp;&nbsp;&nbsp;times:&nbsp;Numpy&nbsp;1-D&nbsp;array&nbsp;of&nbsp;floats<br>
&nbsp;&nbsp;&nbsp;&nbsp;fileName:&nbsp;name&nbsp;of&nbsp;the&nbsp;figure&nbsp;file&nbsp;to&nbsp;save<br>
&nbsp;&nbsp;&nbsp;&nbsp;id:&nbsp;label&nbsp;to&nbsp;add&nbsp;to&nbsp;the&nbsp;figure&nbsp;title<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;None<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#-PlotVisibilityGraph">PlotVisibilityGraph</a>(A,&nbsp;data,&nbsp;times,&nbsp;'FIgure.png',&nbsp;'Test&nbsp;Data')</tt></dd></dl>
 <dl><dt><a name="-addVisibilityGraph"><strong>addVisibilityGraph</strong></a>(data, times, dataName='G1S1', coords=[0.05, 0.95, 0.05, 0.95], numberOfVGs=1, groups_ac_colors=['b'], fig=None, numberOfCommunities=6, printCommunities=False, fontsize=None, nodesize=None, level=0.55, commLineWidth=0.5, lineWidth=1.0, withLabel=True, withTitle=False, layout='circle', radius=0.07, noplot=False)</dt><dd><tt>Draw&nbsp;a&nbsp;Visibility&nbsp;graph&nbsp;of&nbsp;data&nbsp;on&nbsp;a&nbsp;provided&nbsp;Matplotlib&nbsp;figure.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;data:&nbsp;array&nbsp;of&nbsp;data&nbsp;to&nbsp;visualize<br>
&nbsp;&nbsp;&nbsp;&nbsp;times:&nbsp;times&nbsp;corresponding&nbsp;to&nbsp;each&nbsp;data&nbsp;point,&nbsp;used&nbsp;for&nbsp;labels<br>
&nbsp;&nbsp;&nbsp;&nbsp;dataName:&nbsp;label&nbsp;to&nbsp;include&nbsp;in&nbsp;file&nbsp;name<br>
&nbsp;&nbsp;&nbsp;&nbsp;coords:&nbsp;coordinates&nbsp;of&nbsp;location&nbsp;of&nbsp;the&nbsp;plot&nbsp;on&nbsp;the&nbsp;figure<br>
&nbsp;&nbsp;&nbsp;&nbsp;numberOfVGs:&nbsp;number&nbsp;of&nbsp;plots&nbsp;to&nbsp;add&nbsp;to&nbsp;this&nbsp;figure<br>
&nbsp;&nbsp;&nbsp;&nbsp;groups_ac_colors:&nbsp;colors&nbsp;corresponding&nbsp;to&nbsp;different&nbsp;groups&nbsp;of&nbsp;graphs<br>
&nbsp;&nbsp;&nbsp;&nbsp;fig:&nbsp;figure&nbsp;object<br>
&nbsp;&nbsp;&nbsp;&nbsp;printCommunities:&nbsp;print&nbsp;communities&nbsp;details&nbsp;to&nbsp;screen<br>
&nbsp;&nbsp;&nbsp;&nbsp;fontsize:&nbsp;size&nbsp;of&nbsp;labels<br>
&nbsp;&nbsp;&nbsp;&nbsp;nodesize:&nbsp;size&nbsp;of&nbsp;nodes<br>
&nbsp;&nbsp;&nbsp;&nbsp;level:&nbsp;distance&nbsp;of&nbsp;the&nbsp;community&nbsp;lines&nbsp;to&nbsp;nodes<br>
&nbsp;&nbsp;&nbsp;&nbsp;commLineWidth:&nbsp;width&nbsp;of&nbsp;the&nbsp;community&nbsp;lines<br>
&nbsp;&nbsp;&nbsp;&nbsp;lineWidth:&nbsp;width&nbsp;of&nbsp;the&nbsp;edges&nbsp;between&nbsp;nodes<br>
&nbsp;&nbsp;&nbsp;&nbsp;withLabel:&nbsp;include&nbsp;label&nbsp;on&nbsp;plot<br>
&nbsp;&nbsp;&nbsp;&nbsp;withTitle:&nbsp;include&nbsp;title&nbsp;on&nbsp;plot<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;None<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#-addVisibilityGraph">addVisibilityGraph</a>(exampleData,&nbsp;exampleTimes,&nbsp;fig=fig,&nbsp;fontsize=16,&nbsp;nodesize=700,&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;level=0.85,&nbsp;commLineWidth=3.0,&nbsp;lineWidth=2.0,&nbsp;withLabel=False)</tt></dd></dl>
 <dl><dt><a name="-ampSquaredNormed"><strong>ampSquaredNormed</strong></a>(func, freq, times, data)</dt><dd><tt>Lomb-Scargle&nbsp;core&nbsp;function<br>
Calculate&nbsp;the&nbsp;different&nbsp;frequency&nbsp;components&nbsp;of&nbsp;our&nbsp;spectrum:&nbsp;project&nbsp;the&nbsp;cosine/sine&nbsp;component&nbsp;and&nbsp;normalize&nbsp;it:<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;func:&nbsp;Sin&nbsp;or&nbsp;Cos<br>
&nbsp;&nbsp;&nbsp;&nbsp;freq:&nbsp;frequencies&nbsp;(1D&nbsp;array&nbsp;of&nbsp;floats)<br>
&nbsp;&nbsp;&nbsp;&nbsp;times:&nbsp;input&nbsp;times&nbsp;(starting&nbsp;point&nbsp;adjusted&nbsp;w.r.t.dataset&nbsp;times),&nbsp;Zero-padded<br>
&nbsp;&nbsp;&nbsp;&nbsp;data:&nbsp;input&nbsp;Data&nbsp;with&nbsp;the&nbsp;mean&nbsp;subtracted&nbsp;from&nbsp;it,&nbsp;before&nbsp;zero-padding.<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Squared&nbsp;amplitude&nbsp;normalized.<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;coef&nbsp;=&nbsp;<a href="#-ampSquaredNormed">ampSquaredNormed</a>(np.cos,&nbsp;freguency,&nbsp;inputTimesNormed,&nbsp;inputDataCentered)<br>
&nbsp;&nbsp;&nbsp;&nbsp;Intended&nbsp;for&nbsp;internal&nbsp;use&nbsp;only.</tt></dd></dl>
 <dl><dt><a name="-autocorrelation"><strong>autocorrelation</strong></a>(inputTimes, inputData, inputSetTimes, UpperFrequencyFactor=1)</dt><dd><tt>Autocorrelation&nbsp;function<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;inputTimes:&nbsp;times&nbsp;corresponding&nbsp;to&nbsp;provided&nbsp;data&nbsp;points&nbsp;(1D&nbsp;array&nbsp;of&nbsp;floats)<br>
&nbsp;&nbsp;&nbsp;&nbsp;inputData:&nbsp;data&nbsp;points&nbsp;(1D&nbsp;array&nbsp;of&nbsp;floats)<br>
&nbsp;&nbsp;&nbsp;&nbsp;inputSetTimes:&nbsp;a&nbsp;complete&nbsp;set&nbsp;of&nbsp;all&nbsp;possible&nbsp;N&nbsp;times&nbsp;during&nbsp;which&nbsp;data&nbsp;could&nbsp;have&nbsp;been&nbsp;collected.<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Array&nbsp;of&nbsp;time&nbsp;lags&nbsp;with&nbsp;corresponding&nbsp;autocorrelations<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;result&nbsp;=&nbsp;<a href="#-autocorrelation">autocorrelation</a>(inputTimes,&nbsp;inputData,&nbsp;inputSetTimes)</tt></dd></dl>
 <dl><dt><a name="-boxCoxTransform"><strong>boxCoxTransform</strong></a>(subset, lmbda=None, giveLmbda=False)</dt><dd><tt>Power&nbsp;transform&nbsp;from&nbsp;scipy.stats<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;subset:&nbsp;pandas&nbsp;Series.<br>
&nbsp;&nbsp;&nbsp;&nbsp;lmbda:&nbsp;Lambda&nbsp;parameter,&nbsp;if&nbsp;not&nbsp;specified&nbsp;optimal&nbsp;value&nbsp;will&nbsp;be&nbsp;determined<br>
&nbsp;&nbsp;&nbsp;&nbsp;giveLmbda:&nbsp;also&nbsp;return&nbsp;Lambda&nbsp;value<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Transformed&nbsp;subset&nbsp;and&nbsp;Lambda&nbsp;parameter<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;myData&nbsp;=&nbsp;<a href="#-boxCoxTransform">boxCoxTransform</a>(myData)</tt></dd></dl>
 <dl><dt><a name="-boxCoxTransformDataframe"><strong>boxCoxTransformDataframe</strong></a>(df)</dt><dd><tt>Box-cox&nbsp;transform&nbsp;data.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df:&nbsp;pandas&nbsp;DataFrame<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Processed&nbsp;pandas&nbsp;Dataframe<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data&nbsp;=&nbsp;<a href="#-boxCoxTransformDataframe">boxCoxTransformDataframe</a>(df_data)</tt></dd></dl>
 <dl><dt><a name="-chop"><strong>chop</strong></a>(expr, tolerance=1e-10)</dt><dd><tt>Equivalent&nbsp;of&nbsp;Mathematica.Chop&nbsp;Function.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;expr:&nbsp;a&nbsp;number&nbsp;or&nbsp;a&nbsp;pyhton&nbsp;sequence&nbsp;of&nbsp;numbers<br>
&nbsp;&nbsp;&nbsp;&nbsp;tolerance:&nbsp;default&nbsp;is&nbsp;the&nbsp;same&nbsp;as&nbsp;in&nbsp;Mathematica<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Chopped&nbsp;data<br>
&nbsp;<br>
Usage<br>
&nbsp;&nbsp;&nbsp;&nbsp;data&nbsp;=&nbsp;<a href="#-chop">chop</a>(data)</tt></dd></dl>
 <dl><dt><a name="-compareTimeSeriesToPointDataframe"><strong>compareTimeSeriesToPointDataframe</strong></a>(df, point='first')</dt><dd><tt>Subtract&nbsp;a&nbsp;particular&nbsp;point&nbsp;of&nbsp;each&nbsp;time&nbsp;series&nbsp;(row)&nbsp;of&nbsp;a&nbsp;Dataframe.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df:&nbsp;pandas&nbsp;DataFrame<br>
&nbsp;&nbsp;&nbsp;&nbsp;point:&nbsp;'first',&nbsp;'last',&nbsp;0,&nbsp;1,&nbsp;...&nbsp;,&nbsp;10,&nbsp;or&nbsp;a&nbsp;value.<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Processed&nbsp;pandas&nbsp;Dataframe<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data&nbsp;=&nbsp;<a href="#-compareTimeSeriesToPointDataframe">compareTimeSeriesToPointDataframe</a>(df_data)</tt></dd></dl>
 <dl><dt><a name="-compareTwoTimeSeriesDataframe"><strong>compareTwoTimeSeriesDataframe</strong></a>(df1, df2, function=&lt;ufunc 'subtract'&gt;, compareAllLevelsInIndex=True, mergeFunction=&lt;function mean at 0x7fad301bd840&gt;)</dt><dd><tt>Create&nbsp;a&nbsp;new&nbsp;Dataframe&nbsp;based&nbsp;on&nbsp;comparison&nbsp;of&nbsp;two&nbsp;existing&nbsp;Dataframes.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df1:&nbsp;pandas&nbsp;DataFrame<br>
&nbsp;&nbsp;&nbsp;&nbsp;df2:&nbsp;pandas&nbsp;DataFrame<br>
&nbsp;&nbsp;&nbsp;&nbsp;function:&nbsp;np.subtract&nbsp;(default),&nbsp;np.add,&nbsp;np.divide,&nbsp;or&nbsp;another&nbsp;&lt;ufunc&gt;.<br>
&nbsp;&nbsp;&nbsp;&nbsp;compareAllLevelsInIndex:&nbsp;True&nbsp;(default),&nbsp;if&nbsp;False&nbsp;only&nbsp;"source"&nbsp;and&nbsp;"id"&nbsp;will&nbsp;be&nbsp;compared,<br>
&nbsp;&nbsp;&nbsp;&nbsp;mergeFunction:&nbsp;input&nbsp;Dataframes&nbsp;are&nbsp;merged&nbsp;with&nbsp;this&nbsp;function,&nbsp;i.e.&nbsp;np.mean&nbsp;(default),&nbsp;np.median,&nbsp;np.max,&nbsp;or&nbsp;another&nbsp;&lt;ufunc&gt;.<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;New&nbsp;merged&nbsp;pandas&nbsp;Dataframe<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data&nbsp;=&nbsp;<a href="#-compareTwoTimeSeriesDataframe">compareTwoTimeSeriesDataframe</a>(df_dataH2,&nbsp;df_dataH1,&nbsp;function=np.subtract,&nbsp;compareAllLevelsInIndex=False,&nbsp;mergeFunction=np.median)</tt></dd></dl>
 <dl><dt><a name="-createDirectories"><strong>createDirectories</strong></a>(path)</dt><dd><tt>Create&nbsp;a&nbsp;path&nbsp;of&nbsp;directories,&nbsp;unless&nbsp;the&nbsp;path&nbsp;already&nbsp;exists.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;path:&nbsp;path&nbsp;directory<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;None<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#-createDirectories">createDirectories</a>("/pathToFolder1/pathToSubFolder2")</tt></dd></dl>
 <dl><dt><a name="-createReverseDictionary"><strong>createReverseDictionary</strong></a>(inputDictionary)</dt><dd><tt>Efficient&nbsp;way&nbsp;to&nbsp;create&nbsp;a&nbsp;reverse&nbsp;dictionary&nbsp;from&nbsp;a&nbsp;dictionary.<br>
Utilizes&nbsp;Pandas.Dataframe.groupby&nbsp;and&nbsp;Numpy&nbsp;arrays&nbsp;indexing.<br>
&nbsp;<br>
Args:&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;inputDictionary:&nbsp;a&nbsp;dictionary&nbsp;to&nbsp;reverse<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Reversed&nbsp;dictionary<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;revDict&nbsp;=&nbsp;<a href="#-createReverseDictionary">createReverseDictionary</a>(Dict)</tt></dd></dl>
 <dl><dt><a name="-exportClusteringObject"><strong>exportClusteringObject</strong></a>(ClusteringObject, saveDir, dataName, includeData=True, includeAutocorr=True)</dt><dd><tt>Export&nbsp;a&nbsp;clustering&nbsp;Groups-Subgroups&nbsp;dictionary&nbsp;object&nbsp;to&nbsp;a&nbsp;SpreadSheet.<br>
Linkage&nbsp;data&nbsp;is&nbsp;not&nbsp;exported.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;ClusteringObject:&nbsp;clustering&nbsp;object<br>
&nbsp;&nbsp;&nbsp;&nbsp;saveDir:&nbsp;path&nbsp;of&nbsp;directories&nbsp;to&nbsp;save&nbsp;the&nbsp;object&nbsp;to<br>
&nbsp;&nbsp;&nbsp;&nbsp;dataName:&nbsp;label&nbsp;to&nbsp;include&nbsp;in&nbsp;the&nbsp;file&nbsp;name<br>
&nbsp;&nbsp;&nbsp;&nbsp;includeData:&nbsp;export&nbsp;data&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;includeAutocorr:&nbsp;export&nbsp;autocorrelations&nbsp;of&nbsp;data<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;File&nbsp;name&nbsp;of&nbsp;the&nbsp;exported&nbsp;clustering&nbsp;object<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#-exportClusteringObject">exportClusteringObject</a>(myObj,&nbsp;'/dir1',&nbsp;'myObj')</tt></dd></dl>
 <dl><dt><a name="-filterOutAllZeroSignalsDataframe"><strong>filterOutAllZeroSignalsDataframe</strong></a>(df)</dt><dd><tt>Filter&nbsp;out&nbsp;all-zero&nbsp;signals&nbsp;from&nbsp;a&nbsp;DataFrame.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df:&nbsp;pandas&nbsp;DataFrame<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Processed&nbsp;pandas&nbsp;Dataframe<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data&nbsp;=&nbsp;<a href="#-filterOutAllZeroSignalsDataframe">filterOutAllZeroSignalsDataframe</a>(df_data)</tt></dd></dl>
 <dl><dt><a name="-filterOutFirstPointZeroSignalsDataframe"><strong>filterOutFirstPointZeroSignalsDataframe</strong></a>(df)</dt><dd><tt>Filter&nbsp;out&nbsp;out&nbsp;first&nbsp;time&nbsp;point&nbsp;zeros&nbsp;signals&nbsp;from&nbsp;a&nbsp;DataFrame.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df:&nbsp;pandas&nbsp;DataFrame<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Processed&nbsp;pandas&nbsp;Dataframe<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data&nbsp;=&nbsp;<a href="#-filterOutFirstPointZeroSignalsDataframe">filterOutFirstPointZeroSignalsDataframe</a>(df_data)</tt></dd></dl>
 <dl><dt><a name="-filterOutFractionZeroSignalsDataframe"><strong>filterOutFractionZeroSignalsDataframe</strong></a>(df, max_fraction_of_allowed_zeros)</dt><dd><tt>Filter&nbsp;out&nbsp;fraction-zero&nbsp;signals&nbsp;from&nbsp;a&nbsp;DataFrame.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df:&nbsp;pandas&nbsp;DataFrame<br>
&nbsp;&nbsp;&nbsp;&nbsp;max_fraction_of_allowed_zeros:&nbsp;maximum&nbsp;fraction&nbsp;of&nbsp;allowed&nbsp;zeros<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Processed&nbsp;pandas&nbsp;Dataframe<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data&nbsp;=&nbsp;<a href="#-filterOutFractionZeroSignalsDataframe">filterOutFractionZeroSignalsDataframe</a>(df_data,&nbsp;0.75)</tt></dd></dl>
 <dl><dt><a name="-getAdjacencyMatrixOfHorizontalVisibilityGraph"><strong>getAdjacencyMatrixOfHorizontalVisibilityGraph</strong></a>(data)</dt><dd><tt>Calculate&nbsp;adjacency&nbsp;matrix&nbsp;of&nbsp;horizontal&nbsp;visibility&nbsp;graph.<br>
JIT-accelerated&nbsp;version&nbsp;(a&nbsp;bit&nbsp;faster&nbsp;than&nbsp;NumPy-accelerated&nbsp;version).<br>
Single-threaded&nbsp;beats&nbsp;NumPy&nbsp;up&nbsp;to&nbsp;2k&nbsp;data&nbsp;sizes.<br>
Allows&nbsp;use&nbsp;of&nbsp;Multiple&nbsp;CPUs.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;data:&nbsp;Numpy&nbsp;2-D&nbsp;array&nbsp;of&nbsp;floats<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Adjacency&nbsp;matrix<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;A&nbsp;=&nbsp;<a href="#-getAdjacencyMatrixOfHorizontalVisibilityGraph">getAdjacencyMatrixOfHorizontalVisibilityGraph</a>(data)</tt></dd></dl>
 <dl><dt><a name="-getAdjacencyMatrixOfHorizontalVisibilityGraph_NUMPY"><strong>getAdjacencyMatrixOfHorizontalVisibilityGraph_NUMPY</strong></a>(data)</dt><dd><tt>Calculate&nbsp;adjacency&nbsp;matrix&nbsp;of&nbsp;horizontal&nbsp;visibility&nbsp;graph.<br>
NumPy-accelerated&nbsp;version.<br>
Use&nbsp;with&nbsp;datasets&nbsp;larger&nbsp;than&nbsp;2k.<br>
Use&nbsp;in&nbsp;serial&nbsp;applications.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;data:&nbsp;Numpy&nbsp;2-D&nbsp;array&nbsp;of&nbsp;floats<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Adjacency&nbsp;matrix<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;A&nbsp;=&nbsp;<a href="#-getAdjacencyMatrixOfHorizontalVisibilityGraph_NUMPY">getAdjacencyMatrixOfHorizontalVisibilityGraph_NUMPY</a>(data)</tt></dd></dl>
 <dl><dt><a name="-getAdjacencyMatrixOfVisibilityGraph"><strong>getAdjacencyMatrixOfVisibilityGraph</strong></a>(data, times)</dt><dd><tt>Calculate&nbsp;adjacency&nbsp;matrix&nbsp;of&nbsp;visibility&nbsp;graph.<br>
JIT-accelerated&nbsp;version&nbsp;(a&nbsp;bit&nbsp;faster&nbsp;than&nbsp;NumPy-accelerated&nbsp;version).<br>
Allows&nbsp;use&nbsp;of&nbsp;Multiple&nbsp;CPUs.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;data:&nbsp;Numpy&nbsp;2-D&nbsp;array&nbsp;of&nbsp;floats<br>
&nbsp;&nbsp;&nbsp;&nbsp;times:&nbsp;Numpy&nbsp;1-D&nbsp;array&nbsp;of&nbsp;floats<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Adjacency&nbsp;matrix<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;A&nbsp;=&nbsp;getAdjacencyMatrixOfVisibilityGraph_serial(data,&nbsp;times)</tt></dd></dl>
 <dl><dt><a name="-getAdjacencyMatrixOfVisibilityGraph_NUMPY"><strong>getAdjacencyMatrixOfVisibilityGraph_NUMPY</strong></a>(data, times)</dt><dd><tt>Calculate&nbsp;adjacency&nbsp;matrix&nbsp;of&nbsp;visibility&nbsp;graph.<br>
NumPy-accelerated&nbsp;version.&nbsp;Somewhat&nbsp;slower&nbsp;than&nbsp;JIT-accelerated&nbsp;version.<br>
Use&nbsp;in&nbsp;serial&nbsp;applications.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;data:&nbsp;Numpy&nbsp;2-D&nbsp;array&nbsp;of&nbsp;floats<br>
&nbsp;&nbsp;&nbsp;&nbsp;times:&nbsp;Numpy&nbsp;1-D&nbsp;array&nbsp;of&nbsp;floats<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Adjacency&nbsp;matrix<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;A&nbsp;=&nbsp;getAdjacencyMatrixOfVisibilityGraph_serial(data,&nbsp;times)</tt></dd></dl>
 <dl><dt><a name="-getAutocorrelationsOfData"><strong>getAutocorrelationsOfData</strong></a>(params)</dt><dd><tt>Calculate&nbsp;autocorrelation&nbsp;using&nbsp;Lomb-Scargle&nbsp;Autocorrelation.<br>
NOTE:&nbsp;there&nbsp;should&nbsp;be&nbsp;already&nbsp;no&nbsp;missing&nbsp;or&nbsp;non-numeric&nbsp;points&nbsp;in&nbsp;the&nbsp;input&nbsp;Series&nbsp;or&nbsp;Dataframe<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;params:&nbsp;a&nbsp;tuple&nbsp;of&nbsp;parameters&nbsp;in&nbsp;the&nbsp;form&nbsp;(df_data,&nbsp;setAllInputTimes),&nbsp;where<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data&nbsp;is&nbsp;a&nbsp;pandas&nbsp;Series&nbsp;or&nbsp;Dataframe,&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;setAllInputTimes&nbsp;is&nbsp;a&nbsp;complete&nbsp;set&nbsp;of&nbsp;all&nbsp;possible&nbsp;N&nbsp;times&nbsp;during&nbsp;which&nbsp;data&nbsp;could&nbsp;have&nbsp;been&nbsp;collected.<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Array&nbsp;of&nbsp;autocorrelations&nbsp;of&nbsp;data.<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;result&nbsp;&nbsp;=&nbsp;<a href="#-autocorrelation">autocorrelation</a>(df_data,&nbsp;setAllInputTimes)</tt></dd></dl>
 <dl><dt><a name="-getEstimatedNumberOfClusters"><strong>getEstimatedNumberOfClusters</strong></a>(data, cluster_num_min, cluster_num_max, trials_to_do, numberOfAvailableCPUs=4, plotID=None, printScores=False)</dt><dd><tt>Get&nbsp;estimated&nbsp;number&nbsp;of&nbsp;clusters&nbsp;using&nbsp;ARI&nbsp;with&nbsp;KMeans<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;data:&nbsp;data&nbsp;to&nbsp;analyze<br>
&nbsp;&nbsp;&nbsp;&nbsp;cluster_num_min:&nbsp;minimum&nbsp;possible&nbsp;number&nbsp;of&nbsp;clusters<br>
&nbsp;&nbsp;&nbsp;&nbsp;cluster_num_max:&nbsp;maximum&nbsp;possible&nbsp;number&nbsp;of&nbsp;clusters<br>
&nbsp;&nbsp;&nbsp;&nbsp;trials_to_do:&nbsp;number&nbsp;of&nbsp;trials&nbsp;to&nbsp;do&nbsp;in&nbsp;ARI&nbsp;function<br>
&nbsp;&nbsp;&nbsp;&nbsp;numberOfAvailableCPUs:&nbsp;number&nbsp;of&nbsp;processes&nbsp;to&nbsp;run&nbsp;in&nbsp;parallel<br>
&nbsp;&nbsp;&nbsp;&nbsp;plotID:&nbsp;label&nbsp;for&nbsp;the&nbsp;plot&nbsp;of&nbsp;peaks<br>
&nbsp;&nbsp;&nbsp;&nbsp;printScores:&nbsp;print&nbsp;all&nbsp;scores<br>
&nbsp;<br>
Returns:&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;Largest&nbsp;peak,&nbsp;other&nbsp;possible&nbsp;peaks.<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;n_clusters&nbsp;=&nbsp;<a href="#-getEstimatedNumberOfClusters">getEstimatedNumberOfClusters</a>(data,&nbsp;1,&nbsp;20,&nbsp;25)</tt></dd></dl>
 <dl><dt><a name="-getGroupingIndex"><strong>getGroupingIndex</strong></a>(data, n_groups=None, method='weighted', metric='correlation', significance='Elbow')</dt><dd><tt>Cluster&nbsp;data&nbsp;into&nbsp;N&nbsp;groups,&nbsp;if&nbsp;N&nbsp;is&nbsp;provided,&nbsp;else&nbsp;determine&nbsp;N<br>
return:&nbsp;linkage&nbsp;matrix,&nbsp;cluster&nbsp;labels,&nbsp;possible&nbsp;cluster&nbsp;labels.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;data:&nbsp;data&nbsp;to&nbsp;analyze<br>
&nbsp;&nbsp;&nbsp;&nbsp;n_groups:&nbsp;number&nbsp;of&nbsp;groups&nbsp;to&nbsp;split&nbsp;data&nbsp;into<br>
&nbsp;&nbsp;&nbsp;&nbsp;method:&nbsp;linkage&nbsp;calculation&nbsp;method<br>
&nbsp;&nbsp;&nbsp;&nbsp;metric:&nbsp;distance&nbsp;measure<br>
&nbsp;&nbsp;&nbsp;&nbsp;significance:&nbsp;method&nbsp;for&nbsp;determining&nbsp;optimal&nbsp;number&nbsp;of&nbsp;groups&nbsp;and&nbsp;subgroups<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Linkage&nbsp;matrix,&nbsp;cluster&nbsp;index,&nbsp;possible&nbsp;groups<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;x,&nbsp;y,&nbsp;z&nbsp;=&nbsp;<a href="#-getGroupingIndex">getGroupingIndex</a>(data,&nbsp;method='weighted',&nbsp;metric='correlation',&nbsp;significance='Elbow')</tt></dd></dl>
 <dl><dt><a name="-getLobmScarglePeriodogramOfDataframe"><strong>getLobmScarglePeriodogramOfDataframe</strong></a>(df_data, NumberOfCPUs=4, parallel=True)</dt><dd><tt>Calculate&nbsp;Lobm-Scargle&nbsp;periodogram&nbsp;of&nbsp;DataFrame.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df:&nbsp;pandas&nbsp;DataFrame<br>
&nbsp;&nbsp;&nbsp;&nbsp;parallel:&nbsp;calculate&nbsp;in&nbsp;parallel&nbsp;mode&nbsp;(&gt;1&nbsp;process)<br>
&nbsp;&nbsp;&nbsp;&nbsp;NumberOfCPUs:&nbsp;number&nbsp;of&nbsp;processes&nbsp;to&nbsp;create&nbsp;if&nbsp;parallel<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;New&nbsp;pandas&nbsp;Dataframe<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_periodograms&nbsp;=&nbsp;<a href="#-getLobmScarglePeriodogramOfDataframe">getLobmScarglePeriodogramOfDataframe</a>(df_data)</tt></dd></dl>
 <dl><dt><a name="-getRandomAutocorrelations"><strong>getRandomAutocorrelations</strong></a>(df_data, NumberOfRandomSamples=100000, NumberOfCPUs=4)</dt><dd><tt>Generate&nbsp;autocorrelation&nbsp;null-distribution&nbsp;from&nbsp;permutated&nbsp;data&nbsp;using&nbsp;Lomb-Scargle&nbsp;Autocorrelation.<br>
NOTE:&nbsp;there&nbsp;should&nbsp;be&nbsp;already&nbsp;no&nbsp;missing&nbsp;or&nbsp;non-numeric&nbsp;points&nbsp;in&nbsp;the&nbsp;input&nbsp;Series&nbsp;or&nbsp;Dataframe<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data:&nbsp;pandas&nbsp;Series&nbsp;or&nbsp;Dataframe<br>
&nbsp;&nbsp;&nbsp;&nbsp;NumberOfRandomSamples:&nbsp;size&nbsp;of&nbsp;the&nbsp;distribution&nbsp;to&nbsp;generate<br>
&nbsp;&nbsp;&nbsp;&nbsp;NumberOfCPUs:&nbsp;number&nbsp;of&nbsp;processes&nbsp;to&nbsp;run&nbsp;simultaneously<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;DataFrame&nbsp;containing&nbsp;autocorrelations&nbsp;of&nbsp;null-distribution&nbsp;of&nbsp;data.<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;result&nbsp;=&nbsp;<a href="#-getRandomAutocorrelations">getRandomAutocorrelations</a>(df_data)</tt></dd></dl>
 <dl><dt><a name="-getRandomPeriodograms"><strong>getRandomPeriodograms</strong></a>(df_data, NumberOfRandomSamples=100000, NumberOfCPUs=4)</dt><dd><tt>Generate&nbsp;periodograms&nbsp;null-distribution&nbsp;from&nbsp;permutated&nbsp;data&nbsp;using&nbsp;Lomb-Scargle&nbsp;function.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data:&nbsp;pandas&nbsp;Series&nbsp;or&nbsp;Dataframe<br>
&nbsp;&nbsp;&nbsp;&nbsp;NumberOfRandomSamples:&nbsp;size&nbsp;of&nbsp;the&nbsp;distribution&nbsp;to&nbsp;generate<br>
&nbsp;&nbsp;&nbsp;&nbsp;NumberOfCPUs:&nbsp;number&nbsp;of&nbsp;processes&nbsp;to&nbsp;run&nbsp;simultaneously<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;New&nbsp;Pandas&nbsp;DataFrame&nbsp;containing&nbsp;periodograms<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;result&nbsp;=&nbsp;<a href="#-getRandomPeriodograms">getRandomPeriodograms</a>(df_data)</tt></dd></dl>
 <dl><dt><a name="-getSpikes"><strong>getSpikes</strong></a>(inputData, func, cutoffs)</dt><dd><tt>Get&nbsp;sorted&nbsp;index&nbsp;of&nbsp;signals&nbsp;with&nbsp;statistically&nbsp;significant&nbsp;spikes,<br>
i.e.&nbsp;those&nbsp;that&nbsp;pass&nbsp;the&nbsp;provided&nbsp;cutoff.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;inputData:&nbsp;data&nbsp;points&nbsp;(2D&nbsp;array&nbsp;of&nbsp;floats)&nbsp;where&nbsp;rows&nbsp;are&nbsp;normalized&nbsp;signals<br>
&nbsp;&nbsp;&nbsp;&nbsp;func:&nbsp;np.max&nbsp;or&nbsp;np.min<br>
&nbsp;&nbsp;&nbsp;&nbsp;cutoffs:&nbsp;a&nbsp;dictionary&nbsp;of&nbsp;cutoff&nbsp;values<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Index&nbsp;of&nbsp;data&nbsp;with&nbsp;statistically&nbsp;significant&nbsp;spikes<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;index&nbsp;=&nbsp;<a href="#-getSpikes">getSpikes</a>(inputData,&nbsp;np.max,&nbsp;cutoffs)</tt></dd></dl>
 <dl><dt><a name="-getSpikesCutoffs"><strong>getSpikesCutoffs</strong></a>(df_data, p_cutoff, NumberOfRandomSamples=1000)</dt><dd><tt>Calculate&nbsp;spikes&nbsp;cuttoffs&nbsp;from&nbsp;a&nbsp;bootstrap&nbsp;of&nbsp;provided&nbsp;data,<br>
gived&nbsp;the&nbsp;significance&nbsp;cutoff&nbsp;p_cutoff.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data:&nbsp;pandas&nbsp;DataFrame&nbsp;where&nbsp;rows&nbsp;are&nbsp;normalized&nbsp;signals<br>
&nbsp;&nbsp;&nbsp;&nbsp;p_cutoff:&nbsp;p-Value&nbsp;cutoff,&nbsp;e.g.&nbsp;0.01<br>
&nbsp;&nbsp;&nbsp;&nbsp;NumberOfRandomSamples:&nbsp;size&nbsp;of&nbsp;the&nbsp;bootstrap&nbsp;distribution<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Dictionary&nbsp;of&nbsp;spike&nbsp;cutoffs.<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;cutoffs&nbsp;=&nbsp;<a href="#-getSpikesCutoffs">getSpikesCutoffs</a>(df_data,&nbsp;0.01)</tt></dd></dl>
 <dl><dt><a name="-get_optimal_number_clusters_from_linkage_Elbow"><strong>get_optimal_number_clusters_from_linkage_Elbow</strong></a>(Y)</dt><dd><tt>Get&nbsp;optimal&nbsp;number&nbsp;clusters&nbsp;from&nbsp;linkage.<br>
A&nbsp;point&nbsp;of&nbsp;the&nbsp;highest&nbsp;accelleration&nbsp;of&nbsp;the&nbsp;fusion&nbsp;coefficient&nbsp;of&nbsp;the&nbsp;given&nbsp;linkage.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Y:&nbsp;linkage&nbsp;matrix<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Optimal&nbsp;number&nbsp;of&nbsp;clusters<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;n_clusters&nbsp;=&nbsp;<a href="#-get_optimal_number_clusters_from_linkage_Elbow">get_optimal_number_clusters_from_linkage_Elbow</a>(Y)</tt></dd></dl>
 <dl><dt><a name="-get_optimal_number_clusters_from_linkage_Silhouette"><strong>get_optimal_number_clusters_from_linkage_Silhouette</strong></a>(Y, data, metric)</dt><dd><tt>Determine&nbsp;the&nbsp;optimal&nbsp;number&nbsp;of&nbsp;cluster&nbsp;in&nbsp;data&nbsp;maximizing&nbsp;the&nbsp;Silhouette&nbsp;score.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Y:&nbsp;linkage&nbsp;matrix<br>
&nbsp;&nbsp;&nbsp;&nbsp;data:&nbsp;data&nbsp;to&nbsp;analyze<br>
&nbsp;&nbsp;&nbsp;&nbsp;metric:&nbsp;distance&nbsp;measure<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Optimal&nbsp;number&nbsp;of&nbsp;clusters<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;n_clusters&nbsp;=&nbsp;<a href="#-get_optimal_number_clusters_from_linkage_Elbow">get_optimal_number_clusters_from_linkage_Elbow</a>(Y,&nbsp;data,&nbsp;'euclidean')</tt></dd></dl>
 <dl><dt><a name="-hdf5_usage_information"><strong>hdf5_usage_information</strong></a>()</dt><dd><tt>Store/export&nbsp;any&nbsp;lagge&nbsp;datasets&nbsp;in&nbsp;hdf5&nbsp;format&nbsp;via&nbsp;'pandas'&nbsp;or&nbsp;'h5py'<br>
&nbsp;<br>
#&nbsp;mode='w'&nbsp;creates/recreates&nbsp;file&nbsp;from&nbsp;scratch<br>
#&nbsp;mode='a'&nbsp;creates&nbsp;(if&nbsp;no&nbsp;file&nbsp;exists)&nbsp;or&nbsp;appends&nbsp;to&nbsp;the&nbsp;existing&nbsp;file,&nbsp;and&nbsp;reads&nbsp;it<br>
#&nbsp;mode='r'&nbsp;is&nbsp;read&nbsp;only<br>
&nbsp;<br>
#&nbsp;Save&nbsp;data&nbsp;to&nbsp;file&nbsp;using&nbsp;'pandas':&nbsp;<br>
df_example&nbsp;=&nbsp;pd.DataFrame({'A':&nbsp;[1,&nbsp;2,&nbsp;3],&nbsp;'B':&nbsp;[4,&nbsp;5,&nbsp;6]},&nbsp;index=['a',&nbsp;'b',&nbsp;'c'])<br>
df_example.to_hdf('data.h5',&nbsp;key='my_df1',&nbsp;mode='a')<br>
or<br>
series_example&nbsp;=&nbsp;pd.Series([1,&nbsp;2,&nbsp;3,&nbsp;4])<br>
series_example.to_hdf('data.h5',&nbsp;key='my_series',&nbsp;mode='a')<br>
&nbsp;<br>
#&nbsp;Create&nbsp;groups&nbsp;and&nbsp;datasets&nbsp;using&nbsp;'h5py'&nbsp;and&nbsp;'numpy'&nbsp;arrays:<br>
tempFile&nbsp;=&nbsp;h5py.File('data.h5',&nbsp;'a')<br>
tempArray&nbsp;=&nbsp;np.array([[1,2,3,4,5],[6,7,8,9,10]]).astype(float)<br>
if&nbsp;not&nbsp;'arrays/my_array'&nbsp;in&nbsp;tempFile:<br>
&nbsp;&nbsp;&nbsp;&nbsp;dataset_example&nbsp;=&nbsp;tempFile.create_dataset('arrays/my_array',&nbsp;data=tempArray,&nbsp;maxshape=(None,2),&nbsp;dtype=tempArray.dtype,&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;chunks=True)&nbsp;#auto-chunked,&nbsp;else&nbsp;use&nbsp;e.g.&nbsp;chunks=(100,&nbsp;2)<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;#compression='gzip',&nbsp;compression_opts=6<br>
else:<br>
&nbsp;&nbsp;&nbsp;&nbsp;dataset_example&nbsp;=&nbsp;tempFile['arrays/my_array']<br>
&nbsp;<br>
group_example&nbsp;=&nbsp;tempFile.create_group('more_data/additional')<br>
&nbsp;<br>
#&nbsp;Modify&nbsp;values&nbsp;by&nbsp;slicing&nbsp;the&nbsp;dataset&nbsp;or&nbsp;replacing&nbsp;etire&nbsp;one&nbsp;using&nbsp;[...]<br>
dataset_example[:]&nbsp;=&nbsp;np.array([[10,2,3,4,1],[60,7,8,9,1]])<br>
&nbsp;<br>
#&nbsp;New&nbsp;shapes&nbsp;cannot&nbsp;be&nbsp;broadcasted,&nbsp;the&nbsp;dataset&nbsp;needs&nbsp;to&nbsp;be&nbsp;resized&nbsp;explicitly<br>
dataset_example.resize(dataset_example.shape[0]+10,&nbsp;axis=0)&nbsp;#add&nbsp;more&nbsp;rows&nbsp;(initiated&nbsp;with&nbsp;zeros)<br>
&nbsp;<br>
&nbsp;<br>
#&nbsp;Read&nbsp;data&nbsp;from&nbsp;h5&nbsp;file:<br>
df_example&nbsp;=&nbsp;pd.read_hdf('data.h5',&nbsp;'my_df1')<br>
&nbsp;<br>
tempFile&nbsp;=&nbsp;h5py.File('data.h5',&nbsp;'r')<br>
array_example&nbsp;=&nbsp;tempFile['arrays/my_array'].value</tt></dd></dl>
 <dl><dt><a name="-internalAnalysisFunction"><strong>internalAnalysisFunction</strong></a>(data, multiCorr, MultipleList, OutputID, InputID, Species, totalMembers, pValueCutoff, ReportFilterFunction, ReportFilter, TestFunction, HypothesisFunction, FilterSignificant, AssignmentForwardDictionary, AssignmentReverseDictionary, prefix, infoDict)</dt><dd><tt>Analysis&nbsp;for&nbsp;Multi-Omics&nbsp;or&nbsp;Single-Omics&nbsp;input&nbsp;list<br>
The&nbsp;function&nbsp;is&nbsp;used&nbsp;internally&nbsp;and&nbsp;not&nbsp;intended&nbsp;to&nbsp;be&nbsp;used&nbsp;directly&nbsp;by&nbsp;user.<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Intended&nbsp;for&nbsp;internal&nbsp;use</tt></dd></dl>
 <dl><dt><a name="-makeClusteringObject"><strong>makeClusteringObject</strong></a>(df_data, df_data_autocorr, significance='Elbow')</dt><dd><tt>Make&nbsp;a&nbsp;clustering&nbsp;Groups-Subgroups&nbsp;dictionary&nbsp;object.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data:&nbsp;data&nbsp;to&nbsp;analyze&nbsp;in&nbsp;DataFrame&nbsp;format<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data_autocorr:&nbsp;autocorrelations&nbsp;or&nbsp;periodograms&nbsp;in&nbsp;DataFrame&nbsp;format<br>
&nbsp;&nbsp;&nbsp;&nbsp;significance:&nbsp;method&nbsp;for&nbsp;determining&nbsp;optimal&nbsp;number&nbsp;of&nbsp;groups&nbsp;and&nbsp;subgroups<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Clustering&nbsp;object<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;myObj&nbsp;=&nbsp;<a href="#-makeClusteringObject">makeClusteringObject</a>(df_data,&nbsp;df_data_autocorr,&nbsp;significance='Elbow')</tt></dd></dl>
 <dl><dt><a name="-makeDataHistograms"><strong>makeDataHistograms</strong></a>(df, saveDir, dataName)</dt><dd><tt>Make&nbsp;a&nbsp;histogram&nbsp;for&nbsp;each&nbsp;pandas&nbsp;Series&nbsp;(time&nbsp;point)&nbsp;in&nbsp;a&nbsp;pandas&nbsp;Dataframe.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df:&nbsp;DataFrame&nbsp;containing&nbsp;data&nbsp;to&nbsp;visualize<br>
&nbsp;&nbsp;&nbsp;&nbsp;saveDir:&nbsp;path&nbsp;of&nbsp;directories&nbsp;to&nbsp;save&nbsp;the&nbsp;object&nbsp;to<br>
&nbsp;&nbsp;&nbsp;&nbsp;dataName:&nbsp;label&nbsp;to&nbsp;include&nbsp;in&nbsp;the&nbsp;file&nbsp;name<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;None<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#-makeDataHistograms">makeDataHistograms</a>(df,&nbsp;'/dir1',&nbsp;'myData')</tt></dd></dl>
 <dl><dt><a name="-makeDendrogramHeatmap"><strong>makeDendrogramHeatmap</strong></a>(ClusteringObject, saveDir, dataName, AutocorrNotPeriodogr=True, textScale=1.0, vectorImage=True)</dt><dd><tt>Make&nbsp;Dendrogram-Heatmap&nbsp;plot&nbsp;along&nbsp;with&nbsp;VIsibility&nbsp;graphs.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;ClusteringObject:&nbsp;clustering&nbsp;object<br>
&nbsp;&nbsp;&nbsp;&nbsp;saveDir:&nbsp;path&nbsp;of&nbsp;directories&nbsp;to&nbsp;save&nbsp;the&nbsp;object&nbsp;to<br>
&nbsp;&nbsp;&nbsp;&nbsp;dataName:&nbsp;label&nbsp;to&nbsp;include&nbsp;in&nbsp;the&nbsp;file&nbsp;name<br>
&nbsp;&nbsp;&nbsp;&nbsp;AutocorrNotPeriodogr:&nbsp;export&nbsp;data<br>
&nbsp;&nbsp;&nbsp;&nbsp;textScale:&nbsp;scaling&nbsp;of&nbsp;text&nbsp;size<br>
&nbsp;&nbsp;&nbsp;&nbsp;vectorImage:&nbsp;Boolean&nbsp;for&nbsp;exporting&nbsp;vector&nbsp;graphics&nbsp;or&nbsp;PNG&nbsp;format&nbsp;<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;None<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#-makeDendrogramHeatmap">makeDendrogramHeatmap</a>(myObj,&nbsp;'/dir1',&nbsp;'myData',&nbsp;AutocorrNotPeriodogr=True)</tt></dd></dl>
 <dl><dt><a name="-makeLombScarglePeriodograms"><strong>makeLombScarglePeriodograms</strong></a>(df, saveDir, dataName)</dt><dd><tt>Make&nbsp;a&nbsp;combined&nbsp;plot&nbsp;of&nbsp;the&nbsp;signal&nbsp;and&nbsp;its&nbsp;Lomb-Scargle&nbsp;periodogram<br>
for&nbsp;each&nbsp;pandas&nbsp;Series&nbsp;(time&nbsp;point)&nbsp;in&nbsp;a&nbsp;pandas&nbsp;Dataframe.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df:&nbsp;DataFrame&nbsp;containing&nbsp;data&nbsp;to&nbsp;visualize<br>
&nbsp;&nbsp;&nbsp;&nbsp;saveDir:&nbsp;path&nbsp;of&nbsp;directories&nbsp;to&nbsp;save&nbsp;the&nbsp;object&nbsp;to<br>
&nbsp;&nbsp;&nbsp;&nbsp;dataName:&nbsp;label&nbsp;to&nbsp;include&nbsp;in&nbsp;the&nbsp;file&nbsp;name<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;None<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#-makeLombScarglePeriodograms">makeLombScarglePeriodograms</a>(df,&nbsp;'/dir1',&nbsp;'myData')</tt></dd></dl>
 <dl><dt><a name="-mergeDataframes"><strong>mergeDataframes</strong></a>(listOfDataframes)</dt><dd><tt>Merge&nbsp;a&nbsp;list&nbsp;of&nbsp;Dataframes&nbsp;(outer&nbsp;join).<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;listOfDataframes:&nbsp;list&nbsp;of&nbsp;pandas&nbsp;DataFrames<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;New&nbsp;pandas&nbsp;Dataframe<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data&nbsp;=&nbsp;<a href="#-mergeDataframes">mergeDataframes</a>([df_data1,&nbsp;df_data2])</tt></dd></dl>
 <dl><dt><a name="-metricCommonEuclidean"><strong>metricCommonEuclidean</strong></a>(u, v)</dt><dd><tt>Metric&nbsp;to&nbsp;calculate&nbsp;'euclidean'&nbsp;distance&nbsp;between&nbsp;vectors&nbsp;u&nbsp;and&nbsp;v&nbsp;<br>
using&nbsp;only&nbsp;common&nbsp;non-missing&nbsp;points&nbsp;(not&nbsp;NaNs).<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;u:&nbsp;Numpy&nbsp;1-D&nbsp;array<br>
&nbsp;&nbsp;&nbsp;&nbsp;v:&nbsp;Numpy&nbsp;1-D&nbsp;array<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Measure&nbsp;of&nbsp;the&nbsp;distance&nbsp;between&nbsp;u&nbsp;and&nbsp;v<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;dist&nbsp;=&nbsp;<a href="#-metricCommonEuclidean">metricCommonEuclidean</a>(u,v)</tt></dd></dl>
 <dl><dt><a name="-modifiedZScore"><strong>modifiedZScore</strong></a>(subset)</dt><dd><tt>Calculate&nbsp;modified&nbsp;z-score&nbsp;of&nbsp;a&nbsp;1D&nbsp;array&nbsp;based&nbsp;on&nbsp;"Median&nbsp;absolute&nbsp;deviation".<br>
Use&nbsp;on&nbsp;1-D&nbsp;arrays&nbsp;only.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;subset:&nbsp;data&nbsp;to&nbsp;transform<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Transformed&nbsp;subset<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;data&nbsp;=&nbsp;<a href="#-modifiedZScore">modifiedZScore</a>(data)</tt></dd></dl>
 <dl><dt><a name="-modifiedZScoreDataframe"><strong>modifiedZScoreDataframe</strong></a>(df)</dt><dd><tt>Z-score&nbsp;(Median-based)&nbsp;transform&nbsp;data.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df:&nbsp;pandas&nbsp;DataFrame<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Processed&nbsp;pandas&nbsp;Dataframe<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data&nbsp;=&nbsp;<a href="#-modifiedZScoreDataframe">modifiedZScoreDataframe</a>(df_data)</tt></dd></dl>
 <dl><dt><a name="-normalizeSignalsToUnityDataframe"><strong>normalizeSignalsToUnityDataframe</strong></a>(df)</dt><dd><tt>Normalize&nbsp;signals&nbsp;to&nbsp;unity.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df:&nbsp;pandas&nbsp;DataFrame<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Processed&nbsp;pandas&nbsp;Dataframe<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data&nbsp;=&nbsp;<a href="#-normalizeSignalsToUnityDataframe">normalizeSignalsToUnityDataframe</a>(df_data)</tt></dd></dl>
 <dl><dt><a name="-obtainConstantGeneDictionary"><strong>obtainConstantGeneDictionary</strong></a>(GeneDictionary, GetGeneDictionaryOptions, AugmentDictionary)</dt><dd><tt>Obtain&nbsp;gene&nbsp;dictionary&nbsp;-&nbsp;if&nbsp;it&nbsp;exists&nbsp;can&nbsp;either&nbsp;augment&nbsp;with&nbsp;new&nbsp;information&nbsp;or&nbsp;Species&nbsp;or&nbsp;create&nbsp;new,&nbsp;<br>
if&nbsp;not&nbsp;exist&nbsp;then&nbsp;create&nbsp;variable.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;GeneDictionary:&nbsp;an&nbsp;existing&nbsp;variable&nbsp;to&nbsp;use&nbsp;as&nbsp;a&nbsp;gene&nbsp;dictionary&nbsp;in&nbsp;annotations.&nbsp;<br>
If&nbsp;set&nbsp;to&nbsp;None&nbsp;the&nbsp;default&nbsp;ConstantGeneDictionary&nbsp;will&nbsp;be&nbsp;used<br>
&nbsp;&nbsp;&nbsp;&nbsp;GetGeneDictionaryOptions:&nbsp;a&nbsp;list&nbsp;of&nbsp;options&nbsp;that&nbsp;will&nbsp;be&nbsp;passed&nbsp;to&nbsp;this&nbsp;internal&nbsp;GetGeneDictionary&nbsp;function<br>
&nbsp;&nbsp;&nbsp;&nbsp;AugmentDictionary:&nbsp;a&nbsp;choice&nbsp;whether&nbsp;or&nbsp;not&nbsp;to&nbsp;augment&nbsp;the&nbsp;current&nbsp;ConstantGeneDictionary&nbsp;global&nbsp;variable&nbsp;or&nbsp;create&nbsp;a&nbsp;new&nbsp;one<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;None<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#-obtainConstantGeneDictionary">obtainConstantGeneDictionary</a>(None,&nbsp;{},&nbsp;False)</tt></dd></dl>
 <dl><dt><a name="-pAutocorrelation"><strong>pAutocorrelation</strong></a>(args)</dt><dd><tt>Wrapper&nbsp;of&nbsp;Autocorrelation&nbsp;function&nbsp;for&nbsp;use&nbsp;with&nbsp;Multiprocessing.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;args:&nbsp;a&nbsp;tuple&nbsp;of&nbsp;arguments&nbsp;in&nbsp;the&nbsp;form&nbsp;(inputTimes,&nbsp;inputData,&nbsp;inputSetTimes)<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Array&nbsp;of&nbsp;time&nbsp;lags&nbsp;with&nbsp;corresponding&nbsp;autocorrelations<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;result&nbsp;=&nbsp;<a href="#-pAutocorrelation">pAutocorrelation</a>((inputTimes,&nbsp;inputData,&nbsp;inputSetTimes))</tt></dd></dl>
 <dl><dt><a name="-pLombScargle"><strong>pLombScargle</strong></a>(args)</dt><dd><tt>Wrapper&nbsp;of&nbsp;LombScargle&nbsp;function&nbsp;for&nbsp;use&nbsp;with&nbsp;Multiprocessing.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;args:&nbsp;a&nbsp;tuple&nbsp;of&nbsp;arguments&nbsp;in&nbsp;the&nbsp;form&nbsp;(inputTimes,&nbsp;inputData,&nbsp;inputSetTimes)<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Array&nbsp;of&nbsp;frequencies&nbsp;with&nbsp;corresponding&nbsp;intensities<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;result&nbsp;=&nbsp;<a href="#-pLombScargle">pLombScargle</a>((inputTimes,&nbsp;inputData,&nbsp;inputSetTimes))</tt></dd></dl>
 <dl><dt><a name="-prepareDataframe"><strong>prepareDataframe</strong></a>(dataDir, dataFileName, AlltimesFileName)</dt><dd><tt>Make&nbsp;a&nbsp;DataFrame&nbsp;from&nbsp;CSV&nbsp;files.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;dataDir:&nbsp;path&nbsp;of&nbsp;directories&nbsp;pointing&nbsp;to&nbsp;data<br>
&nbsp;&nbsp;&nbsp;&nbsp;dataFileName:&nbsp;file&nbsp;name&nbsp;in&nbsp;dataDir<br>
&nbsp;&nbsp;&nbsp;&nbsp;AlltimesFileName:&nbsp;file&nbsp;name&nbsp;in&nbsp;dataDir<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Pandas&nbsp;Dataframe<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data&nbsp;=&nbsp;<a href="#-prepareDataframe">prepareDataframe</a>(dataDir,&nbsp;dataFileName,&nbsp;AlltimesFileName)<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data.index&nbsp;=&nbsp;pd.MultiIndex.from_tuples([(item.split(':')[1],&nbsp;item.split(':')[0].split('_')[0],<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;('&nbsp;'.join(item.split(':')[0].split('_')[1:]),))&nbsp;for&nbsp;item&nbsp;in&nbsp;df_data.index.values],&nbsp;<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;names=['source',&nbsp;'id',&nbsp;'metadata'])</tt></dd></dl>
 <dl><dt><a name="-quantileNormalizeDataframe"><strong>quantileNormalizeDataframe</strong></a>(df)</dt><dd><tt>Quantile&nbsp;Normalize&nbsp;signals&nbsp;to&nbsp;normal&nbsp;distribution.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df:&nbsp;pandas&nbsp;DataFrame<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Processed&nbsp;pandas&nbsp;Dataframe<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data&nbsp;=&nbsp;<a href="#-quantileNormalizeDataframe">quantileNormalizeDataframe</a>(df_data)</tt></dd></dl>
 <dl><dt><a name="-read"><strong>read</strong></a>(fileName, withPKLZextension=True, hdf5fileName=None, jsonFormat=False)</dt><dd><tt>Read&nbsp;object&nbsp;from&nbsp;a&nbsp;file&nbsp;recorded&nbsp;by&nbsp;function&nbsp;"write".&nbsp;Pandas&nbsp;and&nbsp;Numpy&nbsp;objects&nbsp;are<br>
read&nbsp;from&nbsp;HDF5&nbsp;file&nbsp;when&nbsp;provided,&nbsp;otherwise&nbsp;attempt&nbsp;to&nbsp;read&nbsp;from&nbsp;PKLZ&nbsp;file.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;fileName:&nbsp;path&nbsp;of&nbsp;directories&nbsp;ending&nbsp;with&nbsp;the&nbsp;file&nbsp;name<br>
&nbsp;&nbsp;&nbsp;&nbsp;withPKLZextension:&nbsp;add&nbsp;".pklz"&nbsp;to&nbsp;a&nbsp;pickle&nbsp;file<br>
&nbsp;&nbsp;&nbsp;&nbsp;hdf5fileName:&nbsp;path&nbsp;of&nbsp;directories&nbsp;ending&nbsp;with&nbsp;the&nbsp;file&nbsp;name.&nbsp;If&nbsp;None&nbsp;then&nbsp;data&nbsp;is&nbsp;pickled<br>
&nbsp;&nbsp;&nbsp;&nbsp;jsonFormat:&nbsp;save&nbsp;data&nbsp;into&nbsp;compressed&nbsp;json&nbsp;file&nbsp;<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;data:&nbsp;data&nbsp;object&nbsp;to&nbsp;write&nbsp;into&nbsp;a&nbsp;file<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;exampleDataFrame&nbsp;=&nbsp;<a href="#-read">read</a>('/dir1/exampleDataFrame',&nbsp;hdf5fileName='/dir2/data.h5')</tt></dd></dl>
 <dl><dt><a name="-readMathIOmicaData"><strong>readMathIOmicaData</strong></a>(fileName)</dt><dd><tt>Read&nbsp;text&nbsp;files&nbsp;exported&nbsp;by&nbsp;MathIOmica&nbsp;and&nbsp;convert&nbsp;to&nbsp;Python&nbsp;data<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;fileName:&nbsp;path&nbsp;of&nbsp;directories&nbsp;and&nbsp;name&nbsp;of&nbsp;the&nbsp;file&nbsp;containing&nbsp;data<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Python&nbsp;data<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;data&nbsp;=&nbsp;<a href="#-readMathIOmicaData">readMathIOmicaData</a>("../../MathIOmica/MathIOmica/MathIOmicaData/ExampleData/rnaExample")</tt></dd></dl>
 <dl><dt><a name="-removeConstantSignalsDataframe"><strong>removeConstantSignalsDataframe</strong></a>(df, theta_cutoff)</dt><dd><tt>Remove&nbsp;constant&nbsp;signals.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df:&nbsp;pandas&nbsp;DataFrame<br>
&nbsp;&nbsp;&nbsp;&nbsp;theta_cutoff:&nbsp;parameter&nbsp;for&nbsp;filtering&nbsp;the&nbsp;signals<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Processed&nbsp;pandas&nbsp;Dataframe<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data&nbsp;=&nbsp;<a href="#-removeConstantSignalsDataframe">removeConstantSignalsDataframe</a>(df_data,&nbsp;0.3)</tt></dd></dl>
 <dl><dt><a name="-runCPUs"><strong>runCPUs</strong></a>(NumberOfAvailableCPUs, func, list_of_tuples_of_func_params)</dt><dd><tt>Parallelize&nbsp;function&nbsp;call&nbsp;with&nbsp;multiprocessing.Pool.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;NumberOfAvailableCPUs:&nbsp;number&nbsp;of&nbsp;processes&nbsp;to&nbsp;create<br>
&nbsp;&nbsp;&nbsp;&nbsp;func:&nbsp;function&nbsp;to&nbsp;apply,&nbsp;must&nbsp;take&nbsp;at&nbsp;most&nbsp;one&nbsp;argument<br>
&nbsp;&nbsp;&nbsp;&nbsp;list_of_tuples_of_func_params:&nbsp;function&nbsp;parameters<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Results&nbsp;of&nbsp;func&nbsp;in&nbsp;a&nbsp;numpy&nbsp;array<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;results&nbsp;=&nbsp;<a href="#-runCPUs">runCPUs</a>(4,&nbsp;pAutocorrelation,&nbsp;[(times[i],&nbsp;data[i],&nbsp;allTimes)&nbsp;for&nbsp;i&nbsp;in&nbsp;range(10)])</tt></dd></dl>
 <dl><dt><a name="-runForClusterNum"><strong>runForClusterNum</strong></a>(arguments)</dt><dd><tt>Calculate&nbsp;Adjusted&nbsp;Rand&nbsp;Index&nbsp;of&nbsp;the&nbsp;data&nbsp;for&nbsp;a&nbsp;range&nbsp;of&nbsp;cluster&nbsp;numbers.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;arguments:&nbsp;a&nbsp;tuple&nbsp;of&nbsp;three&nbsp;parameters&nbsp;int&nbsp;the&nbsp;form<br>
&nbsp;&nbsp;&nbsp;&nbsp;(cluster_num,&nbsp;data_array,&nbsp;trials_to_do),&nbsp;where<br>
&nbsp;&nbsp;&nbsp;&nbsp;cluster_num:&nbsp;maximum&nbsp;number&nbsp;of&nbsp;clusters<br>
&nbsp;&nbsp;&nbsp;&nbsp;data_array:&nbsp;data&nbsp;to&nbsp;test<br>
&nbsp;&nbsp;&nbsp;&nbsp;trials_to_do:&nbsp;number&nbsp;of&nbsp;trials&nbsp;for&nbsp;each&nbsp;cluster&nbsp;number<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Numpy&nbsp;array<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;instPool&nbsp;=&nbsp;multiprocessing.Pool(processes&nbsp;=&nbsp;NumberOfAvailableCPUs)<br>
&nbsp;&nbsp;&nbsp;&nbsp;scores&nbsp;=&nbsp;instPool.map(runForClusterNum,&nbsp;[(cluster_num,&nbsp;copy.deepcopy(data),&nbsp;trials_to_do)&nbsp;for&nbsp;cluster_num&nbsp;in&nbsp;range(cluster_num_min,&nbsp;cluster_num_max&nbsp;+&nbsp;1)])<br>
&nbsp;&nbsp;&nbsp;&nbsp;instPool.close()<br>
&nbsp;&nbsp;&nbsp;&nbsp;instPool.join()</tt></dd></dl>
 <dl><dt><a name="-tagLowValuesDataframe"><strong>tagLowValuesDataframe</strong></a>(df, cutoff, replacement)</dt><dd><tt>Tag&nbsp;low&nbsp;values&nbsp;with&nbsp;replacement&nbsp;value.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df:&nbsp;pandas&nbsp;DataFrame<br>
&nbsp;&nbsp;&nbsp;&nbsp;cutoff:&nbsp;values&nbsp;below&nbsp;the&nbsp;"cutoff"&nbsp;are&nbsp;replaced&nbsp;with&nbsp;"replacement"&nbsp;value<br>
&nbsp;&nbsp;&nbsp;&nbsp;replacement:&nbsp;replacement&nbsp;value<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Processed&nbsp;pandas&nbsp;Dataframe<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data&nbsp;=&nbsp;<a href="#-tagLowValuesDataframe">tagLowValuesDataframe</a>(df_data,&nbsp;1.,&nbsp;1.)</tt></dd></dl>
 <dl><dt><a name="-tagMissingValuesDataframe"><strong>tagMissingValuesDataframe</strong></a>(df)</dt><dd><tt>Tag&nbsp;missing&nbsp;(i.e.&nbsp;zero)&nbsp;values&nbsp;with&nbsp;NaN.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df:&nbsp;pandas&nbsp;DataFrame<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;Processed&nbsp;pandas&nbsp;Dataframe<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data&nbsp;=&nbsp;<a href="#-tagMissingValuesDataframe">tagMissingValuesDataframe</a>(df_data)</tt></dd></dl>
 <dl><dt><a name="-timeSeriesClassification"><strong>timeSeriesClassification</strong></a>(df_data, dataName, saveDir, hdf5fileName=None, p_cutoff=0.05, NumberOfRandomSamples=100000, NumberOfCPUs=4, frequencyBasedClassification=False, calculateAutocorrelations=False, calculatePeriodograms=False)</dt><dd><tt>Time&nbsp;series&nbsp;classification.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;df_data:&nbsp;pandas&nbsp;DataFrame<br>
&nbsp;&nbsp;&nbsp;&nbsp;dataName:&nbsp;data&nbsp;name,&nbsp;e.g.&nbsp;"myData_1"<br>
&nbsp;&nbsp;&nbsp;&nbsp;saveDir:&nbsp;path&nbsp;of&nbsp;directories&nbsp;poining&nbsp;to&nbsp;data&nbsp;storage<br>
&nbsp;&nbsp;&nbsp;&nbsp;hdf5fileName:&nbsp;preferred&nbsp;hdf5&nbsp;file&nbsp;name&nbsp;and&nbsp;location<br>
&nbsp;&nbsp;&nbsp;&nbsp;p_cutoff:&nbsp;significance&nbsp;cutoff&nbsp;signals&nbsp;selection<br>
&nbsp;&nbsp;&nbsp;&nbsp;NumberOfRandomSamples:&nbsp;size&nbsp;of&nbsp;the&nbsp;bootstrap&nbsp;distribution&nbsp;to&nbsp;generate<br>
&nbsp;&nbsp;&nbsp;&nbsp;NumberOfCPUs:&nbsp;number&nbsp;of&nbsp;processes&nbsp;allowed&nbsp;to&nbsp;use&nbsp;in&nbsp;calculations<br>
&nbsp;&nbsp;&nbsp;&nbsp;frequencyBasedClassification:&nbsp;whether&nbsp;Autocorrelation&nbsp;of&nbsp;Frequency&nbsp;based<br>
&nbsp;&nbsp;&nbsp;&nbsp;calculateAutocorrelations:&nbsp;whether&nbsp;to&nbsp;recalculate&nbsp;Autocorrelations<br>
&nbsp;&nbsp;&nbsp;&nbsp;calculatePeriodograms:&nbsp;whether&nbsp;to&nbsp;recalculate&nbsp;Periodograms<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;None<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#-timeSeriesClassification">timeSeriesClassification</a>(df_data,&nbsp;dataName,&nbsp;saveDir,&nbsp;NumberOfRandomSamples&nbsp;=&nbsp;10**5,&nbsp;NumberOfCPUs&nbsp;=&nbsp;4,&nbsp;p_cutoff&nbsp;=&nbsp;0.05,&nbsp;frequencyBasedClassification=False)</tt></dd></dl>
 <dl><dt><a name="-visualizeTimeSeriesClassification"><strong>visualizeTimeSeriesClassification</strong></a>(dataName, saveDir, numberOfLagsToDraw=3, hdf5fileName=None, exportClusteringObjects=False, writeClusteringObjectToBinaries=False, AutocorrNotPeriodogr=True, vectorImage=True)</dt><dd><tt>Visualize&nbsp;time&nbsp;series&nbsp;classification.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;dataName:&nbsp;data&nbsp;name<br>
&nbsp;&nbsp;&nbsp;&nbsp;saveDir:&nbsp;path&nbsp;of&nbsp;directories&nbsp;poining&nbsp;to&nbsp;data&nbsp;storage<br>
&nbsp;&nbsp;&nbsp;&nbsp;numberOfLagsToDraw:&nbsp;first&nbsp;top-N&nbsp;lags&nbsp;(or&nbsp;frequencies)&nbsp;to&nbsp;draw<br>
&nbsp;&nbsp;&nbsp;&nbsp;hdf5fileName:&nbsp;HDF5&nbsp;storage&nbsp;path&nbsp;and&nbsp;name<br>
&nbsp;&nbsp;&nbsp;&nbsp;exportClusteringObjects:&nbsp;export&nbsp;clustering&nbsp;objects&nbsp;to&nbsp;xlsx&nbsp;files<br>
&nbsp;&nbsp;&nbsp;&nbsp;writeClusteringObjectToBinaries:&nbsp;export&nbsp;clustering&nbsp;objects&nbsp;to&nbsp;binary&nbsp;(pickle)&nbsp;files<br>
&nbsp;&nbsp;&nbsp;&nbsp;AutocorrNotPeriodogr:&nbsp;label&nbsp;to&nbsp;print&nbsp;on&nbsp;the&nbsp;plots<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;None<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#-visualizeTimeSeriesClassification">visualizeTimeSeriesClassification</a>('myData_1',&nbsp;'/dir1/dir2/',&nbsp;AutocorrNotPeriodogr=True,&nbsp;writeClusteringObjectToBinaries=True)</tt></dd></dl>
 <dl><dt><a name="-write"><strong>write</strong></a>(data, fileName, withPKLZextension=True, hdf5fileName=None, jsonFormat=False)</dt><dd><tt>Write&nbsp;object&nbsp;into&nbsp;a&nbsp;file.&nbsp;Pandas&nbsp;and&nbsp;Numpy&nbsp;objects&nbsp;are&nbsp;recorded&nbsp;in&nbsp;HDF5&nbsp;format<br>
when&nbsp;'hdf5fileName'&nbsp;is&nbsp;provided&nbsp;otherwise&nbsp;pickled&nbsp;into&nbsp;a&nbsp;new&nbsp;file.<br>
&nbsp;<br>
Args:<br>
&nbsp;&nbsp;&nbsp;&nbsp;data:&nbsp;data&nbsp;object&nbsp;to&nbsp;write&nbsp;into&nbsp;a&nbsp;file<br>
&nbsp;&nbsp;&nbsp;&nbsp;fileName:&nbsp;path&nbsp;of&nbsp;directories&nbsp;ending&nbsp;with&nbsp;the&nbsp;file&nbsp;name<br>
&nbsp;&nbsp;&nbsp;&nbsp;withPKLZextension:&nbsp;add&nbsp;".pklz"&nbsp;to&nbsp;a&nbsp;pickle&nbsp;file<br>
&nbsp;&nbsp;&nbsp;&nbsp;hdf5fileName:&nbsp;path&nbsp;of&nbsp;directories&nbsp;ending&nbsp;with&nbsp;the&nbsp;file&nbsp;name.&nbsp;If&nbsp;None&nbsp;then&nbsp;data&nbsp;is&nbsp;pickled.<br>
&nbsp;&nbsp;&nbsp;&nbsp;jsonFormat:&nbsp;save&nbsp;data&nbsp;into&nbsp;compressed&nbsp;json&nbsp;file&nbsp;<br>
&nbsp;<br>
Returns:<br>
&nbsp;&nbsp;&nbsp;&nbsp;None<br>
&nbsp;<br>
Usage:<br>
&nbsp;&nbsp;&nbsp;&nbsp;<a href="#-write">write</a>(exampleDataFrame,&nbsp;'/dir1/exampleDataFrame',&nbsp;hdf5fileName='/dir2/data.h5')</tt></dd></dl>
</td></tr></table><p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#55aa55">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Data</strong></big></font></td></tr>
    
<tr><td bgcolor="#55aa55"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><strong>ConstantGeneDictionary</strong> = None<br>
<strong>ConstantPyIOmicaDataDirectory</strong> = '/Users/user/anaconda3/envs/pyiomicaTest/lib/python3.7/site-packages/pyiomica/data'<br>
<strong>ConstantPyIOmicaExampleVideosDirectory</strong> = '/Users/user/anaconda3/envs/pyiomicaTest/lib/python3.7/site-packages/pyiomica/data/ExampleVideos'<br>
<strong>ConstantPyIOmicaExamplesDirectory</strong> = '/Users/user/anaconda3/envs/pyiomicaTest/lib/python3.7/site-packages/pyiomica/data/ExampleData'<br>
<strong>PackageDirectory</strong> = '/Users/user/anaconda3/envs/pyiomicaTest/lib/python3.7/site-packages/pyiomica'<br>
<strong>UserDataDirectory</strong> = '/Users/user/anaconda3/envs/pyiomicaTest/lib/python3.7/site-packages/pyiomica/data'<br>
<strong>path</strong> = '/Users/user/anaconda3/envs/pyiomicaTest/lib/python3.7/site-packages/pyiomica/data/ExampleVideos'<br>
<strong>readIn</strong> = PosixPath('/Users/user/anaconda3/envs/pyiomica...thon3.7/site-packages/pyiomica/data/__init__.py')</td></tr></table>
</body></html>